---
title: "Klotho QC"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

The purpose of this document is to run some QC on the
Klotho data. In particular, we look for sample swaps
in the RNA seq data.

This workflow uses data from all animals generated by
1.Klotho_Initial_Data_Visualization.Rmd. If it finds
examples of sample swaps, it writes out an alternative
animal information table that can be used in any workflow
as an alternative to the original table.

```{r startup}
rm(list = ls())

library(here)

processed.data.dir <- here("Results", "all")
results.dir <- here("Results", "QC")

all.fun <- list.files(here("Code"), full.names = TRUE)
for(i in 1:length(all.fun)){source(all.fun[i])}
```


```{r libraries, message = FALSE, warning = FALSE, error = FALSE}
needed.libraries <- c("gprofiler2", "DT", "igraph")
load_libraries(needed.libraries)
```


```{r load_data}
count.data <- read.delim(here("Data", "Mouse", "rsem.merged.gene_counts.tsv"))
condensed.table <- read.csv(file.path(processed.data.dir, "mouse.info.csv"))
diff.test <- readRDS(file.path(processed.data.dir, "gene_Effects_of_sex_climb.RDS"))
expr <- readRDS(file.path(processed.data.dir, "processed_data", "Scaled_Expression.RDS"))
```

We investigated the possibility of sample swaps using sex
as an indicator. We used expression from female- and male-
specific genes and compared their expression to the sex
labels specified in the data from the animal room.

The following plot shows a decomposition of a matrix
containing Xist expression and five male-specific genes.
The majority of the animals separate clearly into two 
groups based on gene expression. Three of these animals 
are labeled with the opposite sex in Climb. There are
an additional two animals that do not fall into either
the male or female group. They are both labeled as female
in the climb database.


```{r qc_swaps, fig.width = 7, fig.height = 7}
y.genes <- c("Ddx3y", "Kdm5d", "Eif2s3y", "Uty")
y.id.table <- gconvert(y.genes, organism = "mmusculus", target = "ENSG")
y.gene.id <- y.id.table[,"target"]
names(y.gene.id)  <- y.id.table[,"name"]

x.genes <- "Xist"
x.id.table <- gconvert(x.genes, organism = "mmusculus", target = "ENSG")
x.gene.id <- x.id.table[,"target"]
names(x.gene.id)  <- x.id.table[,"name"]

x.gene.idx <- which(count.data[,"gene_id"] == x.gene.id[1])
x.gene.expr <- as.numeric(count.data[x.gene.idx,3:ncol(count.data)])

y.gene.idx <- which(count.data[,"gene_id"] %in% y.gene.id)
sex.expr <- count.data[c(y.gene.idx, x.gene.idx),3:ncol(count.data)]

#pheatmap(log2(sex.expr+1), cluster_rows = FALSE, cluster_cols = FALSE)
test <- plot.decomp(t(log2(sex.expr+1)), cols = as.numeric(as.factor(condensed.table[,"sex_climb"])))
legend("topright", col = c(2,1), legend = c("Male", "Female"), pch = 16)

min.pc1 <- -0.05; max.pc1 <- 0.05
not.grouped.idx <- intersect(which(test$u[,1] > min.pc1), which(test$u[,1] < max.pc1))
text(test$u[not.grouped.idx,1], test$u[not.grouped.idx,2], 
  labels = gsub("X", "", colnames(sex.expr)[not.grouped.idx]), pos = 4)

mis.male <- intersect(which(test$u[,1] < -0.05), which(condensed.table[,"sex_climb"] == "Male"))
mis.female <- intersect(which(test$u[,1] > 0.05), which(condensed.table[,"sex_climb"] == "Female"))

text(test$u[mis.male,1], test$u[mis.male,2], 
  labels = gsub("X", "", colnames(sex.expr)[mis.male]), pos = 3, offset = 1, col = 2)

text(test$u[mis.female,1], test$u[mis.female,2], 
  labels = gsub("X", "", colnames(sex.expr)[mis.female]), pos = 2, offset = 1, col = 1)

all.mis <- c(not.grouped.idx, mis.male, mis.female)
```

The following plots show Xist expression vs. expression
of several genes on the Y chromosome. The animals 
identified above are labeled.

I hypothesize that animal 78933 is an XO female and that 
animal 83901 is an XXY male. The other three animals 
might have mislabeled RNA-Seq data. 

```{r x_v_y, fig.height = 9, fig.width = 9}
par(mfrow = c(2,2))
for(i in 1:length(y.gene.id)){
  gene.name <- names(y.gene.id)[i]
  gene.idx <- which(count.data[,"gene_id"] == y.gene.id[i])
  gene.expr <- as.numeric(count.data[gene.idx,3:ncol(count.data)])
  
  expr.order <- order(gene.expr)
  climb.sex.col <- as.numeric(as.factor(condensed.table[,"sex_climb"]))

  plot(log2(x.gene.expr+1), log2(gene.expr+1), col = climb.sex.col, pch = 16, xlab = "Xist expression",
    ylab = paste(gene.name, "expression"), main = gene.name)
  legend("topright", col = c(2,1), legend = c("Male", "Female"), pch = 16)
  text(log2(x.gene.expr+1)[all.mis], log2(gene.expr+1)[all.mis], 
    labels = gsub("X", "", colnames(sex.expr)[all.mis]), pos = c(3, 3, 3, 1, 4))
}
```

A subset of these animals were also not validated by genotype.

```{r mis_table}
datatable(condensed.table[all.mis,])
```

```{r topn}
top.n <- 50
```

To further investigate the sex of these animals, we collected
the top `r top.n` genes that were differentially expressed by sex
and performed a decomposition on the matrix. This does not include
any genes on the Y chromosome, as these have been removed.

We still see three misclassified animals and two intermediate
animals, although, appears more male in this setting, which is
not consistent with an XO female. This really doesn't help, does it?

```{r sex_pc, fig.width = 7, fig.height = 7}

#This code should be run with all animals (not an age cohort)
#we select genes that are most differentially expressed by a 
#given factor and use those in the PC to maximize our ability
#to see mis-matched animals based on that factor

top.diff.idx <- order(diff.test$R2, decreasing = TRUE)[1:top.n]
pc <- plot.decomp(t(expr[top.diff.idx,]), cols = as.numeric(as.factor(condensed.table[,"sex_climb"])))
text(pc$u[all.mis,1], pc$u[all.mis,2], colnames(expr)[all.mis], pos = 3)
legend("topright", col = c(2,1), legend = c("Male", "Female"), pch = 16)
```

The next step is to look for obvious swaps. For each of the mismatched
animals is there another mismatched animal for which the computational 
calls for the RNA-Seq sample and the climb calls are perfectly swapped?

I will use the condensed table to match computational calls and climb calls.

```{r swap_fun}
#a function to check whether the computational calls of one animal 
#match the climb calls of another animal
check_pairs <- function(animal1, animal2, sex_col1 = "sex_climb", 
  sex_col2 = "sex_ge", geno_col1 = "climb_geno", 
  geno_col2 = "seq_genotype", check_type = c("undirected", "directed")){
  
  check_type <- check_type[1] #default is undirected
  climb_sex1 <- animal1[sex_col1]
  comp_sex1 <- animal1[sex_col2]
  climb_geno1 <- animal1[geno_col1]
  comp_geno1 <- animal1[geno_col2]
  #climb_age1 <- animal1["age_m"]
  
  climb_sex2 <- animal2[sex_col1]
  comp_sex2 <- animal2[sex_col2]
  climb_geno2 <- animal2[geno_col1]
  comp_geno2 <- animal2[geno_col2]
  #climb_age2 <- animal2["age_m"]

    if(check_type == "undirected"){
        #animal 1 must match animal 2 and vice versa
        match_geno <- (comp_geno1 == climb_geno2) && (comp_geno2 == climb_geno1)
        match_sex <- (comp_sex1 == climb_sex2) && (comp_sex2 == climb_sex1)
    }else{
        #animal 1 must match animal 2, but doesn't need to match the other way
        match_geno <- (comp_geno1 == climb_geno2)
        match_sex <- (comp_sex1 == climb_sex2)
    }
  
  if(match_geno && match_sex){
    result <- 1
  }else{
    result <- 0
  }
  
  return(result)
}
```

```{r swaps}
not.validated <- union(which(condensed.table[,"validated.sex"] == "FALSE"), 
    which(condensed.table[,"validated.gt"] == "FALSE"))

sub.validate <- condensed.table[not.validated,]

animal.pairs <- pair.matrix(1:nrow(sub.validate))
possible.swaps <- rep(0, nrow(sub.validate))

for(i in 1:nrow(animal.pairs)){
    idx1 <- animal.pairs[i,1]
    idx2 <- animal.pairs[i,2]
    possible.swaps[i] <- check_pairs(animal1 = sub.validate[idx1,], animal2 = sub.validate[idx2,])
}

test.pairs <- animal.pairs[which(possible.swaps > 0),]
```

There were no animals among those that were not validated computationally 
that had an obvious swap. 

The next thing to look at is whether there were not individual swaps,
but some kind of rotation of samples. For examples, samples a, b, and
c got put into slots b, c, and a, respectively. To check this, we look
in a directed manner to see if there are any closed loops that for which
the sequencing calls match the climb calls.

The graph below shows which animal sequence calls match the 
climb calls of another animal. There are no loops, so there
is still no obvious way that the samples got mixed up.

```{r check_for_loops}

possible.rotation <- rep(0, nrow(sub.validate))

for(i in 1:nrow(animal.pairs)){
    idx1 <- animal.pairs[i,1]
    idx2 <- animal.pairs[i,2]
    possible.rotation[i] <- check_pairs(animal1 = sub.validate[idx1,], animal2 = sub.validate[idx2,], 
        check_type = "directed")
}

rotation.edges <- animal.pairs[which(possible.rotation > 0),]
swap_graph <- cbind(as.character(sub.validate[rotation.edges[,1], "animalName"]),
    as.character(sub.validate[rotation.edges[,2], "animalName"]))
rotation.net <- graph_from_edgelist(swap_graph)
plot(rotation.net)
```

Part of the difficulty here is that there are multiple
replicates in a single sex-genotype group. If any swaps
happened within a group, we would not be able to see it.
And we can't validate any of the other factors using sequencing,
so we are stuck.

## Conclusion

There are three animals  with mis-matches between their
phenotypic sex and their transcriptional sex. With no way
to identify swaps, we have to assume these are isolated
errors, and we will remove these animals: 100936, 80970,
and 98558.

Two other animals seem to have chromosomal abnormalities.
We think that animal 78933 is an X0 female. This would 
explain low Xist expression and low Y chromosome expression.

We think that the animal 83901 is an XXY male. This would
explain the high Xist expression and high Y chromosome 
expression.

For these two animals we will use their apparent chromosomal
sex. We will leave 78933 as female, and convert 83901 to 
male.

I created a file by hand: Data > general > ID_QC.txt
that contains the IDs of the animals we want to remove
or switch sex.

These will be altered programatically at the beginning 
of 1.Klotho_Initial_Data_Visualization.Rmd.


