---
title: "Klotho interactions with age"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---


## Introduction

The purpose of this workflow is to look at interactions between
age and genotype in affecting gene expression.

```{r param}
rm(list = ls())

library(here)

results.dir <- here("Results", "Age_Effects")
if(!file.exists(results.dir)){
  dir.create(results.dir, TRUE)
  }

geno.cols <- c("WT/WT" = "darkgray", "WT/VS" = "#a6bddb", "VS/VS" = "#2b8cbe", "WT/FC" = "#a1d99b", "FC/FC" = "#31a354")
#barplot(rep(1, length(geno.cols)), col = geno.cols)

general.data.dir <- here("Data", "general")
```


```{r load_code}
all.fun <- list.files(here("Code"), full.names = TRUE)
for(i in 1:length(all.fun)){source(all.fun[i])}
```


```{r libraries, message = FALSE, warning = FALSE, error = FALSE}
needed.libraries <- c("pheatmap", "gprofiler2", "vioplot")
load_libraries(needed.libraries)
```

```{r load_data}
#read data from all ages analysis
adj.expr <- readRDS(here("Results", "all", "processed_data", "Adjusted_gene_Expression_for_climb_geno_test.RDS"))
covar <- read.csv(here("Results", "all", "covariates.csv"))
mouse.gene.table <- read.table(file.path(general.data.dir, "mouse_gene_info.txt"), sep = "\t", header = TRUE)
```

We first compared the variance explained by genotype
at each age for every gene. The boxplot below shows
that genotype explains slightly more variance at 12
months than at 4 months.

```{r var_by_age}
#Look at the variance explained by genotype for each
#transcript in both young and old mice. See if genotype
#explains more variance in the older mice.

u_age <- unique(covar[,"age_batch"])
age.idx <- lapply(u_age, function(x) which(covar[,"age_batch"] == x))
names(age.idx)  <- paste(u_age, "months")

age.test <- lapply(age.idx, function(y) apply(adj.expr[y,], 2, function(x) lm(x~as.factor(covar[y,"climb_geno"]))))
age.r2 <- lapply(age.test, function(y) sapply(y, function(x) summary(x)$adj.r.squared))

boxplot(age.r2, ylab = "Variance Explained by Genotype")
```

The plot below shows that there isn't a very strong relationship
between variance explained at one age and variance explained at
another.

```{r compare}
plot.with.model(age.r2[[1]], age.r2[[2]], report = "cor.test",
    xlab = names(age.r2)[1], ylab = names(age.r2)[2])
```

The plots below show functional enrichments for genes for 
which more variance is explained by genotype at different
ages.

```{r diff_enrich, fig.width = 12, fig.height = 7}

r2.diff <- age.r2[[2]] - age.r2[[1]]
more.in.old <- names(head(sort(r2.diff, decreasing = TRUE), 500))
more.in.young <- names(head(sort(r2.diff, decreasing = FALSE), 500))

more.in.old.enrich <- gost(more.in.old, organism = "mmusculus", 
	sources = c("GO", "KEGG", "REACTOME", "HP", "CORUM", "WP"))
more.in.young.enrich <- gost(more.in.young, organism = "mmusculus", 
	sources = c("GO", "KEGG", "REACTOME", "HP", "CORUM", "WP"))

#quartz(width = 12, height = 7)
par(mfrow = c(1,2))
plot.enrichment(more.in.young.enrich, plot.label = "More genotype-dependent in young mice", 	
	max.term.size = 3000, num.terms = 30)
plot.enrichment(more.in.old.enrich, plot.label = "More genotype-dependent in old mice", 	
	max.term.size = 3000, num.terms = 30)
```

```{r old_young_fun}


young_v_old <- function(gene.name = NULL, tx.id = NULL){

  if(is.null(tx.id)){
    tx.id <- mouse.gene.table[which(mouse.gene.table[,"external_gene_name"] == gene.name),"ensembl_gene_id"]
  }
  if(is.null(gene.name)){
    gene.name <- mouse.gene.table[which(mouse.gene.table[,"ensembl_gene_id"] == tx.id),"external_gene_name"]
    if(length(gene.name) == 0){
      gene.name <- tx.id
    }
  }

  age.expr <- vector(mode = "list", length = length(age.idx))
  names(age.expr) <- names(age.idx)

  for(a in 1:length(age.idx)){
    #quartz()
    age.expr[[a]] <- plot_tx_with_genotype(expr.mat = t(adj.expr[age.idx[[a]],]), 
      covar.table = covar[age.idx[[a]],c(1:2,4)], tx_name = tx.id, 
      ylab = "Count", tx_label = paste(gene.name, "in", names(age.idx)[a], "animals"), 
      pt_col = "#c51b8a", plot.results = TRUE)
  }

  age.mean <- lapply(age.expr, function(x) sapply(x[[1]][[1]], function(y) mean(y)))
  mean.min <- min(unlist(age.mean))
  mean.max <- max(unlist(age.mean))
  
  ybounds <- get_plot_bounds(mean.min, mean.max, scale.factor = 5)
  plot.height = mean.max - mean.min

  #quartz(width = 6, height = 5)
  layout(matrix(c(1,2), nrow = 1, byrow = TRUE), widths = c(1, 0.3))
  par(mar = c(4,4,4,0))
  plot.new()
  plot.window(xlim = c(0.5,2.5), ylim = ybounds)
  for(a in 1:length(age.mean)){
    points(x = rep(a, length(age.mean[[a]])), y = age.mean[[a]], cex = 1.5,
      col = geno.cols[match(names(age.mean[[a]]), names(geno.cols))], pch = 16)
  }
  segments(x0 = rep(1, length(age.mean[[1]])), x1 = rep(2, length(age.mean[[1]])),
    y0 = age.mean[[1]], y1 = age.mean[[2]], lwd = 1.5,
    col = geno.cols[match(names(age.mean[[2]]), names(geno.cols))])
  axis(2, at = round(seq(ybounds[1], ybounds[2], 0.4), 2), las = 2)
  mtext(gene.name, side = 3)
  par(xpd = NA)
  text(x = 1, y = ybounds[1], labels = "4 months")
  text(x = 2, y = ybounds[1], labels = "12 months")
  par(xpd = TRUE)
  par(mar = c(4,0,4,4))
  plot.new()
  plot.window(xlim = c(0,1), ylim = c(0,1))
  legend(x = 0, y = 0.9, legend = names(geno.cols), pch = 16, lty = 1, col = geno.cols)

  quartz()
  interaction.plot(covar[,"age_batch"], covar[,"climb_geno"], adj.expr[,tx.id],
    xlab = "Age", ylab = "Effect", col = geno.cols[match(names(age.mean[[1]]), names(geno.cols))], 
    lwd = 3, trace.label = "genotype", main = gene.name)

    invisible(age.mean)
}


young_old_mean <- function(gene.name = NULL, tx.id = NULL){
  
  if(is.null(tx.id)){
    tx.id <- mouse.gene.table[which(mouse.gene.table[,"external_gene_name"] == gene.name),"ensembl_gene_id"]
  }
  if(is.null(gene.name)){
    gene.name <- mouse.gene.table[which(mouse.gene.table[,"ensembl_gene_id"] == tx.id),"external_gene_name"]
  }

  young.expr <- plot_tx_with_genotype(expr.mat = t(adj.expr[young.idx,]), 
    covar.table = covar[young.idx,c(1:2,4)], tx_name = tx.id, 
    ylab = "Count", tx_label = paste(gene.name, "in young animals"), 
    pt_col = "#c51b8a", plot.results = FALSE)

  
  old.expr <- plot_tx_with_genotype(expr.mat = t(adj.expr[old.idx,]), 
    covar.table = covar[old.idx,c(1:2,4)], tx_name = tx.id, 
    ylab = "Count", tx_label = paste(gene.name, "in old animals"), 
    pt_col = "#c51b8a", plot.results = FALSE)

    young.mean <- sapply(young.expr$grouped_vals[[1]], mean)
    old.mean <- sapply(old.expr$grouped_vals[[1]], mean)

    result <- list("young_mean" = young.mean, "old_mean" = old.mean)
    return(result)
 
}
```

We used a linear model to test the interaction between age
and genotype for every gene. We identified genes that had 
a significant interaction between age and genotype for at
least one of the genotypes.

The heat map below shows the difference in mean expression
between older and younger mice for the nominally significant
transcripts. Blue cells indicate that gene expression went 
down with age in that genotype, and brown cells indicate that 
gene expression went up with age.

Again we see the two major clusters with FC and VS genotypes
opposing each other. Some genes go up in VS and down in FC,
and others do the opposite.

```{r test_ing}
sig.level = 1e-3

geno.fac <- as.factor(covar[,"climb_geno"])
levels(geno.fac) <- c("WT/WT", "WT/FC", "FC/FC", "WT/VS", "VS/VS")

test.int <- apply(adj.expr, 2, function(x) lm(x~as.factor(covar[,"age_batch"])*geno.fac))
test.effect <- t(sapply(test.int, function(x) summary(x)$coefficients[,"Estimate"]))
test.p <- t(sapply(test.int, function(x) summary(x)$coefficients[,"Pr(>|t|)"]))

#the plot below shows the mean expression difference across age
#by genotype for genes with nominally significant interactions
#between age and genotype. The FC/FC genotype seems to have
#the biggest difference in gene expression based on age.

int.col <- 7:10
sig.idx <- apply(test.p[,int.col], 2, function(x) which(x <= sig.level))
sig.tx <- unique(unlist(lapply(sig.idx, names)))

geno.idx <- lapply(age.idx, function(y) lapply(names(geno.cols), function(x) which(covar[y,"climb_geno"] == x)))
tx.mean <- lapply(geno.idx, function(y) sapply(y, function(x) colMeans(adj.expr[x, sig.tx])))

diff.mean <- tx.mean[[2]] - tx.mean[[1]]
colnames(diff.mean) <- names(geno.cols)
row.order <- order(plot.decomp(diff.mean, plot.results = FALSE)$u[,1])
imageWithText(diff.mean[row.order,], show.text = FALSE, split.at.vals = TRUE, 
  col.scale = c("blue", "brown"), grad.dir = "ends", row.names = NULL)
```

Below we plot some of the stronger examples of this age
by genotype interaction.

```{r plots}
#plot old and young expression of genes by name
#gene.name <- "Sc5d"; tx.id <- mouse.gene.table[which(mouse.gene.table[,"external_gene_name"] == gene.name),"ensembl_gene_id"]
#tx.id <- "ENSMUSG00000052139"; gene.name <- mouse.gene.table[which(mouse.gene.table[,"ensembl_gene_id"] == tx.id),"external_gene_name"]

young_v_old(gene.name = "Babam2")

young_v_old(tx.id = sig.tx[row.order][4])
young_v_old(tx.id = sig.tx[rev(row.order)][1])

rnd.tx <- sample(sig.tx, 1)
young_v_old(tx.id = rnd.tx)
```