---
title: "Klotho cell types"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

## Introduction

The purpose of this workflow is to combine differential 
gene expression results from previous workflows with 
cell-type-specfic markers from The Allen Brain Cell
(ABC) Atlas of the mouse brain (PMID: 38092916) to 
try to figure out if there are compositional differences
in the VS and FC brains.

I downloaded the excel file from here:
\url{https://allen-brain-cell-atlas.s3.us-west-2.amazonaws.com/index.html#metadata/WMB-taxonomy/20231215/}
It lists clusters of cells with labels and gene combinations.

```{r param}
rm(list = ls())

library(here)

#subgroup <- NULL
subgroup <- list("age_batch" = 12)
#subgroup <- list("age_batch" = 4, "sex_ge" = "Female") #example with more than one filter

if(is.null(subgroup)){
  results.dir <- here("Results", "all")
}else{
  subgroup.results <- paste(sapply(1:length(subgroup), 
    function(x) paste(names(subgroup)[x], subgroup[[x]], sep = "_")), 
    collapse = "_")
  results.dir <- here("Results", subgroup.results)
}

#comparison.type = "Hom" #compares homozygous FC/FC to VS/VS
#comparison.type = "Het" #compares homozygous WT/FC to WT/VS
comparison.type = "Carrier" #compares all carriers of VS to carriers of FC alleles

geno.cols <- c("VS" = "#2b8cbe", "FC" = "#31a354")

general.data.dir <- here("Data", "general")
```

The subgroup analyzed in this workflow is:
`r paste(names(subgroup), subgroup, sep = " = ")`

```{r load_code}
all.fun <- list.files(here("Code"), full.names = TRUE)
for(i in 1:length(all.fun)){source(all.fun[i])}

processed.data.dir <- file.path(results.dir, "processed_data")
general.processed.data.dir <- here("Results", "Processed_Data")
```

```{r libraries, message = FALSE, warning = FALSE, error = FALSE}
needed.libraries <- c("pheatmap", "readxl")
load_libraries(needed.libraries)
```

```{r info}
#gene information tables
mouse.gene.table <- read.table(file.path(general.data.dir, "mouse_gene_info.txt"), sep = "\t", header = TRUE)
brain_clusters <- read_xlsx(here("Data", "Mouse", "cl.df_CCN202307220.xlsx"), 
    col_names = TRUE, sheet = 1)
adj.expr <- readRDS(file.path(results.dir, 
    paste0("Adjusted_Expression_Separated_by_Group_", comparison.type, ".RDS")))
```


There are multiple levels of organization. We will play with 
them to see which level makes the most sense. For now, we 
will use each row of whichever table we read in as the units,
and we can go up hierarchically as we go.

```{r cluster_expr}
col.headers = c("cluster.markers.combo", "merfish.markers.combo", 
    "cluster.TF.markers.combo", "cluster.markers.combo (within subclass)")
col.idx <- match(col.headers, colnames(brain_clusters))
cluster_expr <- vector(mode = "list", length = nrow(brain_clusters))
names(cluster_expr) <- brain_clusters$cl
for(i in 1:nrow(brain_clusters)){
    cluster_genes <- unique(unlist(strsplit(unlist(brain_clusters[i, col.idx]), ",")))
    gene_id <- mouse.gene.table[match(cluster_genes, mouse.gene.table[,"external_gene_name"]), "ensembl_gene_id"]
    gene.idx <- lapply(adj.expr, function(x) which(names(colnames(x)) %in% gene_id))
    group_expr <- lapply(1:length(gene.idx), function(x) adj.expr[[x]][,gene.idx[[x]]])
    names(group_expr) <- names(adj.expr)
    cluster_expr[[i]] <- group_expr
    #boxplot(cluster_expr[[i]])
}

cluster_test <- lapply(cluster_expr, function(x) t.test(x[[1]], x[[2]]))
cluster_table <- t(sapply(cluster_test, function(x) c(x$estimate, x$p.value)))
#head(cluster_table)
colnames(cluster_table) <- c(names(adj.expr), "p_value")
#plot(cluster_table[,1:2])

```

The following plot shows the number of cell types per class.

```{r subtypes, fig.width = 5, fig.height = 7}
classes <- as.vector(brain_clusters$class_label)
supertypes <- as.vector(brain_clusters$supertype_label)
u_class <- unique(classes)

par(mar = c(4,12,2,2))
barplot(sort(sapply(u_class, function(x) length((which(classes == x))))), las = 2, 
    horiz = TRUE, xlab = "Number of cell types")
```

The following plots show expression differences for different
sets of cells

```{r expr_by_group, fig.width = 10, fig.height = 7}
pdf("~/Desktop/cells.pdf", width = 10, height = 7)
diff.gene.list <- vector(mode = "list", length = length(u_class))
names(diff.gene.list) <- u_class
for(cl in 1:length(u_class)){
    cl.idx <- which(classes == u_class[cl])
    
    sub_group <- cluster_expr[cl.idx]
    names(sub_group) <- supertypes[cl.idx]

    geno.order <- match(names(geno.cols), names(sub_group[[1]]))

    #expression means for each cell type for each group
    expr_means <- lapply(sub_group, function(x) lapply(x, colMeans))
    expr_sd <- lapply(sub_group, function(x) lapply(x, function(y) apply(y, 2, sd)))

   if(length(cl.idx) == 1){
        plot.grouped.boxes(sub_group, plot.grouping = "inner", group.cols = geno.cols[geno.order])        
        next()
    }

    #find mean expression and standard error for each cell type
    mean1 <- sapply(expr_means, function(x) mean(x[[1]]))
    mean2 <- sapply(expr_means, function(x) mean(x[[2]]))

    sd1 <- sapply(expr_means, function(x) sd(x[[1]]))
    sd2 <- sapply(expr_means, function(x) sd(x[[2]]))
    
    group_list <- list(mean1, mean2)
    names(group_list) <- names(cluster_expr[[1]])

    layout.mat <- matrix(c(1,2,3,3), byrow = TRUE, nrow = 2)
    layout(layout.mat)

    par(mar = c(4,4,4,2))
    vioplot(group_list, main = paste(u_class[cl], "\nOverall"), col = geno.cols[geno.order])
    stripchart(group_list, vertical = TRUE, add = TRUE, col = "#c51b8a", method = "jitter", pch = 16)
    
    max.x <- max(c(mean1 + sd1))
    min.x <- min(c(mean1 - sd1))

    max.y <- max(c(mean2 + sd2))
    min.y <- min(c(mean2 - sd2))

    #label points whose error bars don't cross y = x
    no.overlap <- are_equal(x = mean1, x_error = sd1, y = mean2, y_error = sd2)
    u_overlap <- unique(unlist(no.overlap))
    #u_overlap <- no.overlap[[3]]

    if(length(u_overlap) == 0){next()}

    diff.overall <- lapply(group_list, function(x) x[u_overlap])

    par(mar = c(4,4,4,2))
    vioplot(diff.overall, main = paste(u_class[cl], "\nDifferentially Expressed"), col = geno.cols[geno.order])
    stripchart(diff.overall, vertical = TRUE, add = TRUE, col = "#c51b8a", 
        method = "jitter", pch = 16)


    par(mar = c(8, 4, 0, 2))
    no_overlap_list <- expr_means[u_overlap]
    #order by the median of expression in the FC group
    #med.order <- order(sapply(no_overlap_list, function(x) median(x[[1]])), decreasing = FALSE)
    #order by label name
    med.order <- order(names(no_overlap_list))
    plot.grouped.boxes(no_overlap_list[med.order], plot.grouping = "inner", print.vals = NA, 
        label.srt = 90, main = paste(u_class[cl], "Cell Types"), group.cols = geno.cols[geno.order])


    diff.gene.list[[cl]] <- unique(unlist(lapply(no_overlap_list, function(x) names(x[[1]]))))

}
dev.off()
```


```{r diff_genes}
pdf("~/Desktop/diff_decomp.pdf", width = 9, height = 6)
for(cl in 1:length(u_class)){
    par(mfrow = c(2,3))
    class.genes <- diff.gene.list[[cl]]
    class.id <- mouse.gene.table[match(class.genes, mouse.gene.table[,"external_gene_name"]),"ensembl_gene_id"]
    if(length(class.id) < 2){next()}
    gene.idx <- match(class.id, names(colnames(adj.expr[[1]])))
    class_expr <- lapply(adj.expr, function(x) x[,gene.idx])
    names(group_expr) <- names(adj.expr) 
    class_mat <- Reduce("rbind", class_expr)
    
   row.order <- hclust(dist(class_mat))$order
   col.order <- hclust(dist(t(class_mat)))$order
    #pheatmap(class_mat)
    imageWithText(class_mat[row.order,col.order], show.text = FALSE, 
        split.at.vals = TRUE, col.scale = c("blue", "brown"), grad.dir = "ends",
        main = "individual by gene")
    
    pt.col = c(rep("black", nrow(class_expr[[1]])), rep("red", nrow(class_expr[[2]])))
    data.decomp <- plot.decomp(class_mat, cols = pt.col, main = "Decomposition")
    pc.cor <- apply(data.decomp$u, 2, function(x) cor(x, as.numeric(as.factor(pt.col))))
    barplot_with_num(signif(abs(pc.cor), 2), ylab = "Correlation", names = c("PC1", "PC2"), 
        ylim = c(0,0.7), text.shift = 0.1)

    #par(mfrow = c(1,2))
    #cluster_distance(data.decomp$u[,1], data.decomp$u[,2], membership = pt.col, plot.results = TRUE,
    #    xlab = "PC1", ylab = "PC2")
    
    row.order <- hclust(dist(cor(t(class_mat))))$order
    
    imageWithText(cor(t(class_mat))[row.order, row.order], show.text = FALSE, 
        split.at.vals = TRUE, col.scale = c("blue", "brown"), grad.dir = "ends",
        row.names = NULL, col.names = NULL, main = "individual by individual")

    cor.decomp <- plot.decomp(cor(t(class_mat)), cols = pt.col, main = "Decomposition")
    mtext(u_class[cl], side = 3, outer = TRUE, line = -2)
    pc.cor <- apply(cor.decomp$u, 2, function(x) cor(x, as.numeric(as.factor(pt.col))))
    barplot_with_num(signif(abs(pc.cor), 2), ylab = "Correlation", names = c("PC1", "PC2"), 
        ylim = c(0,0.7), text.shift = 0.1)

    #par(mfrow = c(1,2))
    #cluster_distance(cor.decomp$u[,1], cor.decomp$u[,2], membership = pt.col, plot.results = TRUE,
    #    xlab = "PC1", ylab = "PC2")


}
dev.off()
```