---
title: "Klotho differential splicing"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

## Introduction

In previous analyses, we found that the many of the 
differentially expressed genes in the Klotho mice
relate to RNA splicing. 

Dysregulation of splicing has been associated with aging 
(pmid: 31927681). Klotho-reduced mice have been shown to 
have increased intron retention in RNA of multiple organs
(pmid: 34082065). Neurons differentiated from primary human
fibroblasts from aged donors showed depletion of RNA-binding
proteins, and mislocalization of splicing proteins like
TDP-43, which was associated with widespread alternative
splicing. These findings were verified in mouse brains 
(pmid: 40456907).

In this workflow we look specifically at RNA isoforms to identify 
potential variation in isoform production. We 

This workflow depends on results 
generated by 1.Klotho_Initial_Data_Visualization.Rmd
with the argument expression.type == "transcript", and the 
subgroup argument matching the subgroup argument here.

The workflow uses DRIMSeq to look at differential exon 
usage. DRIMSeq tests two groups against each other, so
we specify in the initial parameters which genotypes 
we want to compare.

```{r param}
rm(list = ls())

library(here)

expression.type = "transcript"

args <- commandArgs(trailingOnly=T)
subgroup <- args[2]

if(is.na(subgroup)){
  #subgroup <- "all ages"
  subgroup <- list("age_batch" = 12)
  #subgroup <- list("age_batch" = 4)
  #subgroup <- list("age_batch" = 4, "sex_ge" = "Female") #example with more than one filter
}

if(subgroup == "all ages"){
  results.dir <- here("Results", paste0("all_", expression.type))
}else{
  subgroup.results <- paste(sapply(1:length(subgroup), 
    function(x) paste(names(subgroup)[x], subgroup[[x]], sep = "_")), 
    collapse = "_")
  results.dir <- here("Results", paste(subgroup.results, expression.type, sep = "_"))
}
general.data.dir <- here("Data", "general")

carrier.text <- "carriers" #this is always set to carriers to maximize n


#test by genotoype
#test.factor = "genotype"; ref.geno <- "FC"; alt.geno <- "VS"

#or by high and low expression levels of a gene
test.factor = "expr"; gene.name = "Srsf5"; ref.geno <- "low"; alt.geno  <- "high"

if(test.factor == "genotype"){
  test.text <- paste("We are testing differential exon usage based on Klotho genotype.
  The reference allele is", ref.geno, "and the alternative allele is", alt.geno)
}else{
  test.text <- paste("We are testing differential exon usage based on", ref.geno, "and",
  alt.geno, "expression of", gene.name)
}
```



The subgroup analyzed in this workflow is:
`r paste(names(subgroup), subgroup, sep = " = ")`

`r test.text`

```{r load_code}
all.fun <- list.files(here("Code"), full.names = TRUE)
for(i in 1:length(all.fun)){source(all.fun[i])}

raw.data.dir <- here("Data", "Mouse")

processed.data.dir <- file.path(results.dir, "processed_data")
if(!file.exists(processed.data.dir)){dir.create(processed.data.dir, recursive = TRUE)}

general.processed.data.dir <- here("Results", "Processed_Data")
if(!file.exists(general.processed.data.dir)){dir.create(general.processed.data.dir)}

orig.geno.cols <- get_geno_cols(here("Data", "general", "geno_cols.csv"))
```

```{r libraries, message = FALSE, warning = FALSE, error = FALSE}
needed.libraries <- c("pheatmap", "DRIMSeq", "gprofiler2", "VennDiagram", "igraph")
load_libraries(needed.libraries)
```

```{r load_data}
tx.info <- read.delim(file.path(general.data.dir, "mouse_transcript_info.txt"))
mouse.info <- read.csv(file.path(processed.data.dir, paste0("mouse_info_", carrier.text, ".csv")))
#add X to make all labels characters
rownames(mouse.info) <- paste0("X", rownames(mouse.info))
tx.expr <- read.delim(file.path(raw.data.dir, "rsem.merged.transcript_counts.tsv"))
```

```{r match_expr_covar}
expr.mat <- as.matrix(tx.expr[,3:ncol(tx.expr)])
rownames(expr.mat) <- tx.expr[,1]
common.mouse <- intersect(colnames(expr.mat), rownames(mouse.info))
info.idx <- match(common.mouse, rownames(mouse.info))
expr.idx <- match(common.mouse, colnames(expr.mat))
matched.expr <- expr.mat[,expr.idx]
matched.info <- mouse.info[info.idx,]
```

Here we set up a data frame with the factors for testing.
This could be genotypes of the Klotho alleles, or high and
low expression of a particular gene.

```{r set_test_factor}
if(test.factor == "genotype"){
  carrier.factor <- rep("WT", nrow(matched.info))
  carrier.factor[grep("VS", matched.info[,"ordered_geno"])] <- "VS"
  carrier.factor[grep("FC", matched.info[,"ordered_geno"])] <- "FC"
  geno_factor <- ordered(carrier.factor, levels = c("FC", "WT", "VS"))
  sub.cols <- c("FC" = orig.geno.cols["FC/FC"], "WT" = orig.geno.cols["WT/WT"], "VS" = orig.geno.cols["VS/VS"])
  geno.cols <- sub.cols
  ref.idx <- which(geno_factor == ref.geno)
  alt.idx <- which(geno_factor == alt.geno)
  test.idx <- c(ref.idx, alt.idx)
}

if(test.factor == "expr"){
  gene.expr  <- readRDS(file.path(gsub("transcript", "gene", results.dir), "processed_data", "Batch_Adjusted_Expr.RDS"))
  carrier.factor <- rep("low", nrow(matched.info))
  gene.idx <- which(tx.info[ ,"external_gene_name"] == gene.name)
  gene.id <- tx.info[gene.idx[1],"ensembl_gene_id"]
  tx.idx <- which(colnames(gene.expr) %in% gene.id)

  gene.expr <- gene.expr[match(gsub("X", "", rownames(matched.info)), rownames(gene.expr)), tx.idx]
  low.expr <- which(gene.expr < mean(gene.expr))
  high.expr <- which(gene.expr > mean(gene.expr))
  carrier.factor[high.expr] <- "high"

  geno_factor <- ordered(carrier.factor, levels = c("low", "high"))
  geno.cols <- c("#3182bd", "#a6611a")
  ref.idx <- which(geno_factor == ref.geno)
  alt.idx <- which(geno_factor == alt.geno)
  test.idx <- c(ref.idx, alt.idx)  
}

plotting.df <- data.frame("sequencingBatch" = as.factor(mouse.info[test.idx,"sequencingBatch"]),
  "sex_ge" = as.factor(mouse.info[test.idx,"sex_ge"]),
  "age_batch" = as.numeric(mouse.info[test.idx,"age_batch"]),
  "ordered_geno" = geno_factor[test.idx]) 
rownames(plotting.df) <- rownames(mouse.info)[test.idx]
```

## Load data

```{r exon_usage}
#test.sample <- 1:1000
test.sample <- 1:nrow(matched.expr)

common.tx <- intersect(rownames(matched.expr)[test.sample], tx.info[,"ensembl_transcript_id"])
expr.idx <- match(common.tx, rownames(matched.expr))
info.idx <- match(common.tx, tx.info[,"ensembl_transcript_id"])
tx.table <- data.frame("gene_id" = tx.info[info.idx,"ensembl_gene_id"], 
  "feature_id" = tx.info[info.idx,"ensembl_transcript_id"])

#if(test.factor == "expr"){
#  #adjust for klotho genotype
#  dummy.geno <- dummy_covar(as.matrix(matched.info[,"ordered_geno"]))
#  adj.expr <- t(adjust(t(matched.expr), dummy.geno))
#  #the minimum count is now below zero. makde all counts positive
#  matched.expr <- adj.expr-min(adj.expr)
#}

#test.sample <- 3:ncol(expr.table)
expr.table <- data.frame(feature_id = tx.info[info.idx,"ensembl_transcript_id"], 
  gene_id = tx.info[info.idx,"ensembl_gene_id"],
  matched.expr[expr.idx,])

#remove the levels from the genotypes for some reason
mouse.table <- data.frame(sample_id = rownames(plotting.df), 
  sequencingBatch = plotting.df[,"sequencingBatch"], 
  sex = plotting.df[,"sex_ge"], genotype = as.character(plotting.df[,"ordered_geno"]))

d <- dmDSdata(counts = expr.table, samples = mouse.table)
#head(counts(d), 3)
#head(samples(d), 3)
```

We filtered out genes and transcripts with low counts.
The following plot shows the final distribution of transcripts per gene.

```{r filter}
d <- dmFilter(d, min_samps_gene_expr = 10, min_samps_feature_expr = 3,
  min_gene_expr = 10, min_feature_expr = 10)
plotData(d)
```


```{r fit_model}
design_full <- model.matrix(~ genotype + sequencingBatch, data = samples(d))

#the stats can take a long time, so we'll save a file
result.file <- file.path(results.dir, paste0("Exon_Usage_Result_", ref.geno, "_vs_", alt.geno, ".RDS"))

if(!file.exists(result.file)){
  d <- dmPrecision(d, design = design_full, BPPARAM = BiocParallel::MulticoreParam())
  #d
  #head(mean_expression(d), 3)
  #common_precision(d)
  #head(genewise_precision(d))
  
  d <- dmFit(d, design = design_full, verbose = 1, BPPARAM = BiocParallel::MulticoreParam())
  #head(proportions(d))
  #head(coefficients(d))
  
  d <- dmTest(d, coef = colnames(design_full)[2], verbose = 1, BPPARAM = BiocParallel::MulticoreParam())
  #head(results(d))
  saveRDS(d, result.file)
}else{
  d <- readRDS(result.file)
}
```

## Precision plot

The following plot shows the relationship between the mean
expression of each transcript and the estimated precision.

```{r precision}
plotPrecision(d)
```

## P value distributions

The following plot shows the qq plot of the p values.
They are quite inflated (which maybe we expect?)

```{r pdist}
qqunif.plot(results(d)$adj_pvalue)
```

The following plots show the distributions of the
p values for transcripts.

```{r pdist_tx}
plotPValues(d)
```

The following plots show the distributions of the
p values for genes.

```{r pdist_gene}
plotPValues(d, level = "feature")
```

```{r results}
res <- results(d)
res <- res[order(res$pvalue, decreasing = FALSE),]
```

## Test significance

Next we use StageR to screen for significant genes.

```{r gene_level, message = FALSE, warning = FALSE, error = FALSE}
library("stageR") #load this here, otherwise it collides with the other packages

## Assign gene-level pvalues to the screening stage
pScreen <- results(d)$pvalue
names(pScreen) <- results(d)$gene_id

## Assign transcript-level pvalues to the confirmation stage
pConfirmation <- matrix(results(d, level = "feature")$pvalue, ncol = 1)
rownames(pConfirmation) <- results(d, level = "feature")$feature_id

## Create the gene-transcript mapping
tx2gene <- results(d, level = "feature")[, c("feature_id", "gene_id")]

## Create the stageRTx object and perform the stage-wise analysis
stageRObj <- stageRTx(pScreen = pScreen, pConfirmation = pConfirmation,
  pScreenAdjusted = FALSE, tx2gene = tx2gene)

stageRObj <- stageWiseAdjustment(object = stageRObj, method = "dtu", 
  alpha = 0.05, allowNA = TRUE)

sig.gene.stager <- getSignificantGenes(stageRObj)
sig.tx <- getSignificantTx(stageRObj)

padj <- getAdjustedPValues(stageRObj, order = TRUE, onlySignificantGenes = FALSE)
#head(padj)
```

```{r effect_size, eval = FALSE}
# Extract fitted proportions
gene.name <- "Stxbp5"
gene.id <- which(tx.gene.info[,"external_gene_name"])
ref.id <- rownames(plotting.df)[which(plotting.df$ordered_geno == ref.geno)]
alt.id <- rownames(plotting.df)[which(plotting.df$ordered_geno == alt.geno)]
props <- proportions(d)
u_gene <- unique(props[,"gene_id"])
for(i in 1:length(u_gene)){
  gene.idx <- which(props[,"gene_id"] == u_gene[i])
  tr.ref.props <- props[gene.idx,ref.id]
  tr.alt.props <- props[gene.idx,alt.id]
  z.ref.prop <- apply(tr.ref.props, 1, scale)
  z.alt.prop <- apply(tr.alt.props, 1, scale)
  ref.mean <- rowMeans(tr.ref.props)  
  alt.mean <- rowMeans(tr.alt.props)  
  barplot(rbind(ref.mean, alt.mean), beside = TRUE)
  plot.grouped.boxes(list("ref" = z.ref.prop, "alt" = z.alt.prop), type = "matrix")
}
```


```{r select_sig}
#alpha <- 1e-6
alpha <- 0.05
sig.gene.idx <- which(padj[,"gene"] <= alpha)
sig.genes <- unique(padj[sig.gene.idx,"geneID"])
#cat(sig.genes, sep = "\n")
sig.tx.idx <- which(padj[,"transcript"] <= alpha)
sig.tx <- unique(padj[sig.tx.idx,"transcript"])

write.table(padj, file.path(results.dir, 
  paste0("DEx_results_", ref.geno, "_vs_", alt.geno, ".csv")), sep = ",",
  quote = FALSE, row.names = FALSE)
```

The plot below shows the distribution of the adjusted p values.
We took all genes with an adjusted p value < 0.05

```{r adj_dist}
u_gene <- unique(padj[,"geneID"])
gene.idx <- match(u_gene, padj[,"geneID"])
qqunif.plot(padj[gene.idx,"gene"], plot.label = "Adjusted p values of genes")
abline(h = -log10(alpha))
```

## Enrichments

There were `r length(sig.genes)` differentially spliced
genes and `r length(sig.tx)` differentially used exons
at an adjusted p value of `r alpha`.

The plot below shows enrichments for alternatively spliced
genes. The alternative splicing seems to be primarily 
related to synaptic genes.

```{r enrich, fig.height = 7, fig.width = 7}
sig.enrich <- gost(sig.genes, organism = "mmusculus", sources = c("GO", "KEGG", "REACTOME", "CORUM", "HP"))
plot.enrichment(sig.enrich, plot.label = "Enrichment of alternatively spliced genes",
  num.terms = 30, max.term.size = 3000)
```

We also plotted out GO term networks showing how the significant
genes related to different GO categories (cellular compartment, 
molecular function, and biological process)


```{r go_parts}
sig.gene.table <- gconvert(sig.genes, organism = "mmusculus")
all.sig.gene.names <- sig.gene.table[,"name"]

detailed.enrich <- gost(all.sig.gene.names, organism = "mmusculus", sources = "GO", evcodes = TRUE)
go.parts <- c("GO:CC", "GO:MF", "GO:BP")

#hist(-log10(as.numeric(detailed.enrich[[1]]$"p_value")))

pdf(file.path(results.dir, "GO_gene_net.pdf"), width = 15, height = 15)
    gene.net <- go_gene_net(enrichment.result = detailed.enrich, max.term.size = 3000,
        plot.results = TRUE, min.vertex.size = 2, max.vertex.size = 8, max.pval = 1e-50)
        mtext("Significant Differential Exon Usage", side = 3, outer = TRUE, line = -2.5)    
dev.off()

for(i in 1:length(go.parts)){
    pdf(file.path(results.dir, paste0(gsub(":", "_", go.parts[i]), "_net.pdf")), 
      width = 15, height = 10)
    sig.net <- go_gene_net(enrichment.result = detailed.enrich, max.term.size = 3000, 
        filter.source = go.parts[i], max.pval = 1e-50,
        plot.results = TRUE, min.vertex.size = 2, max.vertex.size = 8)
        mtext("Significantly Differential Exon Usage", side = 3, outer = TRUE, line = -2.5)    
    
    dev.off()
}
```

## Examples {.tabset .tabset-fade .tabset-pills}

```{r count_ex}
n.examples <- min(c(10, length(sig.genes)))
```

Below we show alternative exon usage for the top `r n.examples` 
most significant genes. Some of them really do look differentially
spliced, but not all are convincing.

```{r examples, results = "asis", fig.height = 7, fig.width = 7}
sig.gene.names <- tx.info[match(sig.genes[1:n.examples], 
  tx.info[,"ensembl_gene_id"]), "external_gene_name"]

if(length(n.examples) > 0){
  for(i in 1:n.examples){
    top_gene_id <- sig.genes[i]
    top_gene_name <- sig.gene.names[i]
    cat("###", top_gene_name, "\n")

    print(plotProportions(d, gene_id = top_gene_id, group_variable = "genotype", 
      plot_type = "boxplot1", group_colors = geno.cols[c(1,3)]))
    
    #plotProportions(d, gene_id = top_gene_id, group_variable = "genotype")
    #plotProportions(d, gene_id = top_gene_id, group_variable = "genotype", plot_type = "lineplot")
    #plotProportions(d, gene_id = top_gene_id, group_variable = "genotype", plot_type = "ribbonplot")
    #plotProportions(d, gene_id = top_gene_id, group_variable = "genotype", plot_type = "boxplot1")
    #plotProportions(d, gene_id = top_gene_id, group_variable = "genotype", plot_type = "boxplot2")
    cat("\n\n")
  }
}
```

```{r spec_gene, eval = FALSE}
gene.name = "Synj1"

pdf(paste0("~/Desktop/", gene.name, ".pdf"), width = 7, height = 7)
gene.id <- tx.info[match(gene.name, tx.info[,"external_gene_name"]), "ensembl_gene_id"]
plotProportions(d, gene_id = gene.id, group_variable = "genotype", 
  plot_type = "boxplot1", group_colors = geno.cols[c(1,3)])
dev.off()
```

## Comparison

If we have run all pairwise genotype comparisons, 
the next code compares the significantly differentially
spliced genes across runs using the above alpha of `r alpha`

There is surprisingly little overlap across the 
comparisons. Maybe this is too noisy an analysis 
to do a comparison across these groups. We have
the most power to detect differences in the VS-FC
comparison, so we will stick with that one.


```{r compare, fig.width = 5, fig.height = 5}
results.files <- list.files(results.dir, pattern = "DEx_results", full.names = TRUE)
if(length(results.files) == 3){
  all.dex <- lapply(results.files, function(x) read.csv(x))
  names(all.dex) <- gsub(".csv", "", gsub("DEx_results_", "", basename(results.files)))
  
  gene.lists <- lapply(all.dex, function(x) unique(x[which(x[,"gene"] <= alpha),"geneID"]))
  plotVenn(gene.lists)
}
```

The bar plot below shows more clearly that the VS-FC comparison
has the largest number of differentially spliced genes. The 
VS-WT comparison has the next most, and the FC-WT comparison
has the fewest.

```{r counts}
if(length(results.files) == 3){
  barplot_with_num(sapply(gene.lists, length), main = "Significant genes in each comparison",
    ylab = "Gene Count")
}
```

The enrichments below show that all comparisons are enriched
for synaptic functions, but the the FC-VS comparison has the 
strongest enrichments. Probably because it is the largest 
group of genes.

```{r comp_enrich, fig.width = 7, fig.heigh = 7}
if(length(results.files) == 3){
  comp.enrich <- lapply(gene.lists, function(x) gost(x, organism = "mmusculus", 
    sources = c("GO", "KEGG")))

  plot.enrichment.group(comp.enrich, transformation = "sqrt", 
    plot.label = "Square root of the -log10(p) for enrichments",
    max.term.size = 3000, cluster_cols = FALSE)
}
```

## Comparison to differentially expressed genes

Do differential expression and differential splicing occur
in the same genes? About a third of the genes with significant
differential exon usage were also differentially expressed.

```{r deg_deu_comparison}
deg.dir <- gsub("transcript", "gene", results.dir)
deg.table <- read.delim(file.path(deg.dir, "DEG_clusters_carriers.txt"))
sig.overlap <- intersect(rownames(deg.table), sig.genes)
plotVenn(list("DEG" = rownames(deg.table), "DEU" = sig.genes))
```

There were `r length(sig.overlap)` genes with differential
expression that also had differential splicing.

The histogram below shows the differential splicing p values
for the genes that were also significantly differentially 
expressed.

```{r overlap_sig}
overlap.p <- padj[which(padj[,"geneID"] %in% sig.overlap),]
hist(-log10(overlap.p[,"gene"]), xlab = "-log10(p)", main = "-log10(p) of overlapping genes")
```

The following enrichment plot shows that these genes are
also enriched for synaptic terms.

```{r overlap_enrich, fig.height = 7, fig.width = 7}
overlap.enrich <- gost(sig.overlap, organism = "mmusculus", sources = "GO")
plot.enrichment(overlap.enrich, max.term.size = 3000, num.terms = 30)
```

## Comparison to previous publications {.tabset .tabset-fade .tabset-pills}

Ravi wrote a paper looking at splicing variation in Trem2 and Apoe 
mice (pmid: 37016304). He identified a small list of high-confidence 
genes that were differentially spliced in both the mice and the human 
data (Table 7). Here we do a quick check to see if any of these genes
are in our list (probably because our list is huge).
It looks as if I'll need to go directly to the other sources if I 
want to get the gene lists for the humans.

The -log10 p values for the overlapping genes are shown below. 
They are not particularly impressive relative to the overall
significance we see in this data set.

```{r pandey_overlap}
pandey <- read.csv(here("Data", "Mouse", "Pandey_et_al_Table7.csv"))
overlap <- intersect(pandey[,"ENSEMBL"], sig.genes)
if(length(overlap) > 0){
  overlap.names <- pandey[match(overlap, pandey[,"ENSEMBL"]), "SYMBOL"]
  overlap.idx <- lapply(overlap, function(x) which(padj[,"geneID"] == x))
  names(overlap.idx) <- overlap.names
  overlap.p <- lapply(overlap.idx, function(x) padj[x,])
  barplot(sapply(overlap.p, function(x) -log10(x[1,"gene"])), ylab = "-log10(p value)")
}else{
  plot.text("No overlap with Pandey genes")
}
```

The following plots show the differential transcript usage
for each of the above genes. Not terribly compelling.

```{r details, results = "asis", fig.height = 7, fig.width = 7}
if(length(overlap) > 0){
  for(i in 1:length(overlap)){
    top_gene_id <- overlap[i]
    top_gene_name <- pandey[which(pandey[,"ENSEMBL"] == top_gene_id),"SYMBOL"]
    cat("###", top_gene_name, "\n")

    print(plotProportions(d, gene_id = top_gene_id, group_variable = "genotype", 
      plot_type = "boxplot1", group_colors = geno.cols[c(1,3)]))
    
    cat("\n\n")
  }
}
```

Otherwise this paper is a huge mess. Every single subdivision of 
the animals you can possible imagine has it's own list of differential
expression and differential exon use from multiple different packages.
There is absolutely no use in looking at overlaps with these gene lists.

## Comparison to Srsf5 connectors

Are the genes that are connected to Srsf5 in the gene interaction network
differentially spliced? No. The line shows a p value of 0.05.

```{r srsf5_net}
#pdf("~/Desktop/Srsf5_conn_deu.pdf")
srsf5.conn <- c("Capzb", "Rpl27a", "Ash1l", "Bsg", "Per3", "Suclg1", "Snrpd1")
conn.id <- tx.info[match(srsf5.conn, tx.info[,"external_gene_name"]), "ensembl_gene_id"]
conn.p <- padj[match(conn.id, padj[,"geneID"]), "gene"]
names(conn.p) <- srsf5.conn
barplot(-log10(conn.p), 
  main = "-log10(p) for differential exon usage\nof genes connected to Srsf5",
  ylim = c(0, -log10(0.01)))
abline(h = -log10(0.05), lty = 2)
```

