---
title: "Klotho human comparison"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

The purpose of this document is to compare differential 
expression across Klotho genotypes to differential expression
in the human AD cohorts. There are three: ROS-MAP, Mayo, and 
Mount Sainai. Each of these has multiple brain regions.
Sage has done the pre-processing to look for differential
expression between AD and control, explained here:
\url{https://www.synapse.org/Synapse:syn23573590/wiki/617653}

We could look at each cohort and each brain region and 
compare to each Klotho genotype to see how well the mouse
differential expression recapitulates what is seen between
AD and control expression. 

We could use the following procedure:

In humans: Test each gene for differential expression in AD and
controls (already done by Sage). Group into biologically 
functional groups. Calculate the difference in mean expression 
for each group using controls as a reference, such that a negative 
number means that the expression is down in AD patients.

In mice: Test each gene for differential expression comparing 
each variant carrier to the WT mice (done in Klotho.Rmd). Group 
into biologically functional groups. Calculate the difference 
in mean expression for each group using WT mice as the reference. 
So a negative number means that the expression is reduced in FC or 
VS carriers.

For comparison: Plot the human difference for each group on the
x axis and the mouse difference on the y axis. 

Do this process for the FC and VS genotypes in mice, and compare
to each brain region in each human cohort.

```{r param}
rm(list = ls())

library(here)

#expression.type = "transcript"
expression.type = "gene"

#subgroup <- NULL
subgroup <- list("age_batch" = 12)
#subgroup <- list("age_batch" = 4, "sex_ge" = "Female") #example with more than one filter

if(is.null(subgroup)){
  results.dir <- here("Results", "all")
}else{
  subgroup.results <- paste(sapply(1:length(subgroup), 
    function(x) paste(names(subgroup)[x], subgroup[[x]], sep = "_")), 
    collapse = "_")
  results.dir <- here("Results", subgroup.results)
}

processed.data.dir <- file.path(results.dir, "processed_data")
```

```{r load_code}
all.fun <- list.files(here("Code"), full.names = TRUE)
for(i in 1:length(all.fun)){source(all.fun[i])}
```

```{r libraries}
needed.libraries <- c("synapser", "pheatmap", "DESeq2", "DT", "vioplot", "RColorBrewer",
  "gprofiler2", "cluster", "pathview", "ClusterProfiler")
load_libraries(needed.libraries)
```

We downloaded the AD-Control comparisons for each cohort from Synapse.
\url{https://www.synapse.org/Synapse:syn26720676}

```{r download_human_data}
de.synID <- c("MSBB" = "syn27068762", "Mayo" = "syn27024969", "ROSMAP" = "syn26967457")
human.data.dir <- here("Data", "Human")
if(!file.exists(human.data.file)){dir.create(human.data.dir)}

if(!logged.in){
    synLogin()
    logged.in <- TRUE
}

for(i in 1:nrow(data_id)){
    dest.file <- list.files(human.data.file, pattern = names(de.synID)[i])
    if(length(dest.file) == 0){
    synGet(as.vector(de.synID[i]), downloadLocation = human.data.dir)
  }
}
```

