---
title: "Klotho human comparison"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

The purpose of this document is to download differential 
in the human AD cohorts. There are three: ROS-MAP, Mayo, and 
Mount Sainai. Each of these has multiple brain regions.
Sage has done the pre-processing to look for differential
expression between AD and control, explained here:
\url{https://www.synapse.org/Synapse:syn23573590/wiki/617653}

We could look at each cohort and each brain region and 
compare to each Klotho genotype to see how well the mouse
differential expression recapitulates what is seen between
AD and control expression. 

We could use the following procedure:

In humans: 

1. Test each gene for differential expression in AD and
  controls (already done by Sage). 
2. Group into biologically functional groups (GO terms, KEGG, etc.).

In mice: 

1. Test each gene for differential expression comparing 
  the VS-carrying mice to the FC-carrying mice (done in 
  1b.Direct_Variant_Comparison.Rmd). 
2. Group mouse genes into biologically functional groups 
  (GO terms, KEGG, etc.). (done in 
  1.Klotho_Initial_Data_Visualization.Rmd)
3. Calculate the difference in mean expression for each group 
  using FC mice as the reference. (done in 
  1b.Direct_Variant_Comparison.Rmd) The FC allele in humans is the 
  common allele. The VS allele confers protection from AD for people 
  carrying the APOE4 allele.

To compare mouse and human:

For each functional group (GO, KEGG, etc.) in each brain region:

1. Find the human mouse orthologs for direct comparison
2. Plot the human fold change on the x axis and the mouse 
  difference on the y axis. 
3. This gives us a score that tells us how AD-like the mouse 
  expression is to the human AD expression in each human brain 
  region and each functional term (GO, KEGG, etc.).


```{r param}
rm(list = ls())

library(here)

#expression.type = "transcript"
expression.type = "gene"

#subgroup <- NULL
subgroup <- list("age_batch" = 12)
#subgroup <- list("age_batch" = 4, "sex_ge" = "Female") #example with more than one filter

if(is.null(subgroup)){
  results.dir <- here("Results", "all")
}else{
  subgroup.results <- paste(sapply(1:length(subgroup), 
    function(x) paste(names(subgroup)[x], subgroup[[x]], sep = "_")), 
    collapse = "_")
  results.dir <- here("Results", subgroup.results)
}


  #comparison.type = "Hom" #compares homozygous FC/FC to VS/VS
  #comparison.type = "Het" #compares homozygous WT/FC to WT/VS
  comparison.type = "Carrier" #compares all carriers of VS to carriers of FC alleles


if(comparison.type == "Hom"){
    vs.allele <- "VS/VS"; fc.allele = "FC/FC"
}
if(comparison.type == "Het"){
    vs.allele <- "WT/VS"; fc.allele = "WT/FC"
}
if(comparison.type == "Carrier"){
    vs.allele <- "VS"; fc.allele = "FC"
}

processed.data.dir <- file.path(results.dir, "processed_data")
```



```{r load_code}
all.fun <- list.files(here("Code"), full.names = TRUE)
for(i in 1:length(all.fun)){source(all.fun[i])}
```

```{r libraries}
needed.libraries <- c("synapser", "pheatmap", "DESeq2", "DT", "vioplot", "RColorBrewer",
  "gprofiler2", "cluster", "pathview", "clusterProfiler")
load_libraries(needed.libraries)
```

We downloaded the AD-Control comparisons for each cohort from Synapse.
\url{https://www.synapse.org/Synapse:syn26720676}

```{r download_human_data}
de.synID <- c("MSBB" = "syn27068762", "Mayo" = "syn27024969", "ROSMAP" = "syn26967457")
human.data.dir <- here("Data", "Human")
if(!file.exists(human.data.dir)){dir.create(human.data.dir)}

logged.in = FALSE

if(!logged.in){
    synLogin()
    logged.in <- TRUE
}

for(i in 1:length(de.synID)){
    dest.file <- list.files(human.data.dir, pattern = names(de.synID)[i])
    if(length(dest.file) == 0){
    synGet(as.vector(de.synID[i]), downloadLocation = human.data.dir)
  }
}
```

```{r read_data}
human.deg.files <- list.files(human.data.dir, pattern = "diagnosis", full.names = TRUE)
human.deg <- lapply(human.deg.files, read.delim)

mouse.deg.file <- paste0(gsub("/", ".", fc.allele), "_vs_", gsub("/", ".", vs.allele), "_DEG.csv")
mouse_deg <- read.csv(file.path(results.dir, mouse.deg.file))
```

```{r parse_table}
#This function groups differential expression results into 
#brain regions

parse_table <- function(de.table){
  u_comp <- unique(de.table$Comparison)
  comp.list <- vector(mode = "list", length = length(u_comp))
  names(comp.list) <- u_comp
}
```

