## Batches Pre-Normalization {.tabset .tabset-fade .tabset-pills}


```{r pc_fun}
plot_pc_factors <- function(expr.decomp){
  for(i in 2:ncol(info)){
    cat("###", colnames(info)[i], "\n")
    pc1.var.exp <- signif(expr.decomp$var.exp[1]*100, 2)
    pc2.var.exp <- signif(expr.decomp$var.exp[2]*100, 2)
    
    layout(matrix(c(1,2), nrow = 1), widths = c(1, 0.4))
    par(mar = c(6,6,4,0))
    if(is.factor(info[,i])){
      plot(expr.decomp$u, pch = 16, xlab = paste0("PC1 (", pc1.var.exp, "%)"),
        ylab = paste0("PC2 (", pc2.var.exp, "%)"), col = as.numeric(info[,i]),
        main = colnames(info)[i])
    }else{
      age.cols = brewer.pal(9, "RdPu")
      u_age <- sort(unique(info[,i]))
      age.col <- sapply(info[,i], function(x) age.cols[which(u_age == x)])
      plot(expr.decomp$u, pch = 16, xlab = paste0("PC1 (", pc1.var.exp, "%)"),
        ylab = paste0("PC2 (", pc2.var.exp, "%)"), col = age.col, main = colnames(info)[i])
    }
    par(mar = c(6,0,4,2))
    plot.new()
    plot.window(xlim = c(0,1), ylim = c(0,1))
    if(is.factor(info[,i])){
      legend(0, 1, legend = levels(info[,i]), pch = 16, col = unique(as.numeric(info[,1])))
    }else{
      legend(0, 1, legend = u_age, pch = 16, col = age.cols[1:length(u_age)])
    }
    cat("\n\n")
  }
}
```

The following plots show the relationships between the first
two PCs of the un-normalized gene expression matrix and relevant
variables.

```{r pre_norm_pc, fig.width = 6, fig.height = 5, results = "asis"}
#pdf("~/Desktop/test.pdf", width = 6, height = 5)
plot_pc_factors(pre.norm.decomp)
#dev.off()
```

## Parsing biodomains


```{r parse_bd}
#make a list of the biodomains. Each biodomain has a subdomain
#with a list of human genes.

annotated_bd <- readRDS(file.path(biodomain.dir, "annotated_biodomains_Oct23.rds"))
u_bd <- unique(annotated_bd$Biodomain)
u_sbd <- unique(annotated_bd$Subdomain)

bd_list <- vector(mode = "list", length = length(u_bd))
names(bd_list) <- u_bd

for(bd in 1:length(u_bd)){
  bd.idx <- which(annotated_bd$Biodomain == u_bd[bd])
  u_sub.bd <- unique(annotated_bd$Subdomain[bd.idx])
  u_sub.bd[which(is.na(u_sub.bd))] <- u_bd[bd]
  sub.list <- vector(mode = "list", length = length(u_sub.bd))
  names(sub.list) <- u_sub.bd
  for(sbd in 1:length(u_sub.bd)){
    if(u_sub.bd[sbd] == u_bd[bd]){
      sbd.idx <- which(is.na(annotated_bd$Subdomain[bd.idx]))
    }else{
      sbd.idx <- which(annotated_bd$Subdomain[bd.idx] == u_sub.bd[sbd])
    }
    sbd.genes <- Reduce("union", annotated_bd$ensembl_id[bd.idx[sbd.idx]])
  }

}
```

From 1.Klotho_Initial_Data_Visualization.Rmd after biodomain
expression section:

## Cluster Enrichment {.tabset .tabset-fade .tabset-pills}

I thought it might be interesting to look at the enrichment
of the obvious clusters in the plots above, but the enrichments
are really redundant, so I'm taking out the plots. It will 
probably be more informative to look at intersections with 
GO terms and KEGG pathways.


```{r cluster_enrich, results = "asis", fig.height = 7, fig.width = 7}
stop()
for(i in 1:length(gene.enrich)){
  cat("###", names(mean.bd.vals)[i], "\n")
  plot.enrichment.group(gene.enrich[[i]], max.term.size = 3000, 
    n.terms = 30, cluster_cols = FALSE, plot.label = names(mean.bd.vals)[i])
  cat("\n\n")
}
```


## from KEGG and Biodomain Intersections


To calculate a significance threshold on the largest contributing
pathways, we need to perform permutations and calculate a null 
distribution for something. What we are interested in is the 
expression's ability to discriminate the genotypes, so we need
to calculate a statistic about that. We shuffled the mouse labels,
which breaks the correlation between genotype and expression,
but not the correlation structure of gene expression.

We need a statistic that quantifies the ability of the gene 
expression to distinguish genotypes along the first principal
component.

```{r perm_fun}
#This function permutes the output of merge_mean_mats()
#This is a matrix with pathway intersections in rows and individuals
#in columns. It shuffles the individual labels and then recalculates
#pathway means for each genotype based on the shuffled individuals.
#it returns the first two principal components of the shuffled matrices.

perm_mean_mat <- function(ind.mean.mat, n.perm = 1000){
    perm.genotype.score <- vector(mode = "list", length = n.perm)

    for(p in 1:n.perm){
        perm.order <- sample(1:ncol(ind.mean.mat)) #shuffle individuals
        
        #shuffle the individual labels on the mean gene expression
        perm.ind.mean.mat <- ind.mean.mat[,perm.order]
        colnames(perm.ind.mean.mat) <- colnames(sub.vals)

        perm.geno.mats <- lapply(geno.idx, function(x) perm.ind.mean.mat[,x])
        perm.geno.means <- sapply(perm.geno.mats, rowMeans)
        
        #plot.decomp(t(perm.geno.means), label.points = TRUE, cols = geno.cols, cex = 1.5)
        mean.perm.decomp <- plot.decomp(t(perm.geno.means), plot.results = FALSE) 
        perm.genotype.score[[p]] <- mean.perm.decomp$u
    }

    return(perm.genotype.score)
}

```

```{r perm}

true.decomp <- plot.decomp(t(sub.vals), plot.results = FALSE)

nperm = 1000
perm.geno.score <- perm_mean_mat(sub.vals, nperm)
```

The following plot shows the score for each genotype along
PC1 against the permuted result for PC1. The VS genotypes
do tend to be on the low end and the FC genotypes tend to
be on the high end, but they are still well within the 
random distributions.

```{r plot_perm_pc1}
perm1 <- t(sapply(perm.geno.score, function(x) x[,1]))
vioplot(perm1, col = geno.cols)
points(x = 1:length(geno.cols), y = mean.decomp$u[,1], pch = "*", cex = 3, col = "#c51b8a")
abline(h = 0)
```

The following plot is the same as above but for PC2

```{r plot_perm_pc2}
perm2 <- t(sapply(perm.geno.score, function(x) x[,2]))
vioplot(perm2, col = geno.cols)
points(x = 1:length(geno.cols), y = mean.decomp$u[,2], pch = "*", cex = 3, col = "#c51b8a")
abline(h = 0)
```

The following plot shows the calculated PC values for each genotype
against random distrbutions of PC1 and PC2. These are not particularly
inspiring. The VS genotypes are at the edge of their distributions, but 
still within the random point cloud.


```{r both_pc, fig.width = 9, fig.height = 6}
par(mfrow = c(2,3))
for(i in 1:ncol(perm1)){
    xmin <- min(c(perm1[,i], mean.decomp$u[i,1]))
    xmax <- max(c(perm1[,i], mean.decomp$u[i,1]))
    ymin <- min(c(perm2[,i], mean.decomp$u[i,2]))
    ymax <- max(c(perm2[,i], mean.decomp$u[i,2]))

    plot(perm1[,i], perm2[,i], xlim = c(xmin, xmax), ylim = c(ymin, ymax),
        xlab = "PC1", ylab = "PC2", main = names(geno.cols)[i],)
    points(mean.decomp$u[i,1], mean.decomp$u[i,2], col = geno.cols[i], pch = 16, cex = 3)
}

```

Maybe the individual scores don't matter? Maybe it's about how well
they are separated or clustered? The VS genotype should group together,
the FC genotypes should group together, and they should be separated 
from each other.

Or maybe we should test individual pathways, not overall?

```{r sandbox, eval = FALSE}
path.name <- "Ribosome"
path.name <- "Tryptophan metabolism"


path.idx <- which(names(mean.bd.k.vals) == path.name)
path.decomp <- plot_mean_bd_k(mean.bd.k.vals[[path.idx]], covar.table)
path.sub <- merge_mean_mats(mean.bd.k.vals[path.idx])

geno.path <- lapply(geno.idx, function(x) path.sub[,x])
geno.path.means <- sapply(geno.path, rowMeans)
mean.path.decomp <- plot.decomp(t(geno.path.means), cols = geno.cols, label.points = TRUE)

perm.path <- perm_mean_mat(path.sub, 1000)

perm1 <- t(sapply(perm.path, function(x) x[,1]))
vioplot(perm1, col = geno.cols, main = path.name)
points(x = 1:length(geno.cols), y = mean.path.decomp$u[,1], pch = "*", cex = 3, col = "#c51b8a")
abline(h = 0)


```


## ANOVA

What if we just do an ANOVA on each intersection.
We've done ANOVA's on individual genes, this will
look at the means of biodomain-kegg intersections.


```{r anova}

test_pathways <- function(nested.gene.list, expr.mat){

    all.aov <- max.mean.dist <- vector(mode = "list", length = length(nested.gene.list))
    names(all.aov) <- names(max.mean.dist) <- names(nested.gene.list)
    
    for(bd in 1:length(nested.gene.list)){
        bd.aov <- max.bd.dist <- vector(mode = "list", length = length(nested.gene.list[[bd]]))
        names(bd.aov) <- names(max.bd.dist) <- names(nested.gene.list[[bd]])
        if(length(nested.gene.list[[bd]]) == 0){next()}
        for(k in 1:length(nested.gene.list[[bd]])){
            gene.ids <- nested.gene.list[[bd]][[k]]
            common.genes <- intersect(gene.ids, colnames(expr.mat))
            gene.mat <- expr.mat[,common.genes]
            geno.mats <- lapply(geno.idx, function(x) t(gene.mat)[,x])
            geno.means <- sapply(geno.mats, rowMeans)
            #boxplot(geno.means)
            bd.aov[[k]] <- aov.by.matrix(geno.means)
            max.bd.dist[[k]] <- as.vector(dist(t(geno.means)))
        }
        all.aov[[bd]] <- bd.aov
        max.mean.dist[[bd]] <- max.bd.dist
    }
    return(list("aov" = all.aov, "dist" = max.mean.dist))
}

get_nested_stats <- function(anova.list, dist.list){
    bd.p <- bd.dist <- matrix(NA, ncol = length(kegg.bd.list), nrow = length(kegg.list))
    colnames(bd.p) <- colnames(bd.dist) <- names(kegg.bd.list)
    rownames(bd.p) <- rownames(bd.dist) <- names(kegg.list)
    for(bd in 1:length(kegg.bd.list)){
        sub.p <- sapply(anova.list[[bd]], function(x) summary(x)[[1]][,"Pr(>F)"][1])
        sub.dist <- sapply(dist.list[[bd]], max)
        if(length(sub.p) > 0){
            bd.p[names(sub.p),bd] <- sub.p
            bd.dist[names(sub.dist),bd] <- sub.dist
        }
    }

    result <- list("p" = bd.p, "dist" = bd.dist)
}


bd.kegg.test <- test_pathways(kegg.bd.list, adj_expr)
bd.kegg.aov <- bd.kegg.test$aov
bd.kegg.dist <- bd.kegg.test$dist

all.dist <- unlist(bd.kegg.dist)
hist(all.dist)

nperm = 10
perm.p <- perm.dist <- vector(mode = "list", length = nperm)
for(p in 1:nperm){
    print(p)
    rnd.sample <- sample(1:nrow(adj_expr))
    sampled.expr <- adj_expr[rnd.sample,]
    rownames(sampled.expr) <- rownames(adj_expr)
    bd.kegg.perm <- test_pathways(kegg.bd.list, sampled.expr)
    
    bd.kegg.aov.perm <- bd.kegg.perm$aov
    bd.kegg.dist.perm <- bd.kegg.perm$dist

    aov.stats <- get_nested_stats(anova.list = bd.kegg.aov.perm, dist.list =  bd.kegg.dist.perm)

    perm.p[[p]] <- aov.stats$p
    perm.dist[[p]] <- aov.stats$dist
}

qqunif.plot(as.vector(unlist(perm.p)))
hist(as.vector(unlist(perm.dist)))

pv <- as.vector(bd.p)
pv <- pv[which(!is.na(pv))]
adj.p <- p.adjust(pv, "fdr")
#length(which(adj.p < 0.05))
p.order <- order(pv, decreasing = FALSE)
test <- cbind("p" = pv[p.order], "adjusted_p" = adj.p[p.order])
thresh.p <- max(test[which(test[,"adjusted_p"] <= 0.00011),"p"])

sig.mat <- bd.p
sig.mat[which(sig.mat > thresh.p)] <- NA
row.vals <- which(apply(sig.mat, 1, function(x) !all(is.na(x))))
col.vals <- which(apply(sig.mat, 2, function(x) !all(is.na(x))))

sig.p <- -log10(bd.p[row.vals, col.vals])
sig.dist <- bd.dist[row.vals, col.vals]


pdf("~/Desktop/sig_intersections.pdf", width = 7, height = 17)
pheatmap(sig.p, cluster_rows = FALSE, cluster_cols = FALSE)
pheatmap(sig.dist, cluster_rows = FALSE, cluster_cols = FALSE)
dev.off()
 
pass.thresh <- which(!is.na(sig.p), arr.ind = TRUE)
diff.names <- apply(pass.thresh, 1, function(x) paste(rownames(sig.p)[x[1]], colnames(sig.p)[x[2]], sep = " - "))
diff.idx <- match(diff.names, rownames(sub.vals))

sub.means <- sapply(geno.idx, function(x) rowMeans(sub.vals[,x]))
pheatmap(sub.means[diff.idx,])

plot_mean_bd_k(sub.vals[diff.idx,], covar.table)
```

Is is interesting that the WT are not intermediate, but span 
the full range? That is that Klotho nudges you to one side of
normal or another side, but doesn't push you outside of the 
normal range.

The WT only seem intermediate because their mean is zero.



## AD gene enrichment stats

Originally, I had a wanted to show that the DEG in the Klotho
mice were enriched for "AD genes." I did this by getting a 
list of AD genes and doing a hypergeometric test to see if
this list were enriched in the DEG. The list I used was the
Agora list, which Greg said might be too controversial to use.

He also suggested that a Venn diagram might not be the most
effective way of showing the enrichment. He suggested instead
going back to the AD-gene heatmap I had at the very beginning
where I had a short-list of genes shown to be genetically 
associated with AD risk. I showed that a number of them, including
APOE and APP were differentially expressed across the Klotho
genotypes. 

So I'm taking out the enrichment stastics and Venn diagrams 
and replacing them with a heatmap. Greg suggested that we can
even show individual stats in the heatmap by having a colum for
each mouse with delimiters, rather than showing means for each
group.

The following plot shows how the differentially expressed
genes and AD-associated genes interact with each other.

```{r venn_fig}
sig.list <- list("All genes" = rownames(effect.fdr), "DEG" = rownames(mean.sig.expr),
  "AD Genes" = ad.ensembl)
#plotVenn(Vlist = sig.list)

circle.cols <- c("all" = "black", "deg" = "#bf812d", "ad" = "#756bb1")

ind.num <- sapply(sig.list, length)
deg.only <- length(setdiff(sig.list$DEG, sig.list$"AD Genes"))
ad.only <- length(setdiff(sig.list$"AD Genes", sig.list$DEG))
deg.and.ad <- length(intersect(sig.list$"AD Genes", sig.list$DEG))

par(mar = c(0,0,0,0))
plot.new()
plot.window(xlim = c(0, 1), ylim = c(0,1))

inner.center <- 0.45
inner.label.y <- 0.62
deg.center <- 0.4 
ad.center <- 0.7
label.cex <- 1.2

#all.circle <- get_circle(radius = 0.45, center_x = 0.5, center_y = inner.center)
#points(all.circle$x, all.circle$y, col = circle.cols["all"], type = "l", lwd = 3)
draw.rectangle(min.x = 0.12, max.x = 0.92, min.y = 0.15, max.y = 0.85, border = circle.cols["all"], lwd = 3)
deg.circle <- get_circle(radius = 0.25, center_x = 0.4, center_y = inner.center)
points(deg.circle$x, deg.circle$y, col = circle.cols["deg"], type = "l", lwd = 3)
ad.circle <- get_circle(radius = 0.15, center_x = 0.7, center_y = inner.center)
points(ad.circle$x, ad.circle$y, col = circle.cols["ad"], type = "l", lwd = 3)


text(x = 0.52, y = 0.81, labels = paste0("All (", ind.num["All genes"], ")"), 
  col = circle.cols["all"], font = 2, cex = label.cex)
text(x = deg.center, y = inner.label.y, labels = paste0("DEG (", ind.num["DEG"], ")"), 
  col = circle.cols["deg"], font = 2, cex = label.cex)
text(x = ad.center+0.035, y = inner.label.y, labels = paste0("AD (", ind.num["AD Genes"], ")"), 
  col = circle.cols["ad"], font = 2, cex = label.cex)

text(x = deg.center, y = inner.center, labels = deg.only, font = 2, cex = label.cex, col = circle.cols["deg"])
text(x = ad.center+0.02, y = inner.center, labels = ad.only, font = 2, cex = label.cex, col = circle.cols["ad"])
text(x = 0.6, y = inner.center, labels = deg.and.ad, font = 2, cex = label.cex, col = mix_colors(circle.cols["deg"], circle.cols["ad"]))
```


## AD gene overview

Are AD genes enriched among the differentially expressed genes?
Are there more AD genes among the DEG than we would expect by 
chance? We used the hypergeometric distribution to answer this.

We have an urn with $N$ total marbles. Some of these marbles
($K$ red marbles) are differentially expressed. $N-K$ green
marbles are not differentially expressed. If we draw $n$ AD 
genes, what is the probability that exaclty $k$ of them are 
red (i.e. differentially expressed)?

. We want to know
the probability of drawing $k$    

```{r hyper, fig.width = 5, fig.height = 5}
#quartz(width = 5, height = 5)
plot.new()
par(mar = c(1,1,1,1))
plot.window(xlim = c(0, 1), ylim = c(0, 1))
draw.rectangle(0.1, 0.9, 0.1, 0.9)

gene.circle <- get_circle(radius = 0.25, center_x = 0.4, center_y = 0.5)
lower.idx <- which(gene.circle$y <= 0.5)
upper.idx <- which(gene.circle$y >= 0.5)
plot.poly.xy(gene.circle$x[upper.idx], gene.circle$y[upper.idx], 
  gene.circle$x[lower.idx], gene.circle$y[lower.idx], border = NA,
  col = rgb(190/256,174/256,212/256))
points(gene.circle$x, gene.circle$y, type = "l")

ad.circle <- get_circle(radius = 0.25, center_x = 0.6, center_y = 0.5)
lower.idx <- which(ad.circle$y <= 0.5)
upper.idx <- which(ad.circle$y >= 0.5)
plot.poly.xy(ad.circle$x[upper.idx], ad.circle$y[upper.idx], 
  ad.circle$x[lower.idx], ad.circle$y[lower.idx], border = NA,
  col = rgb(127/256,201/256,127/256, alpha = 0.5))
points(ad.circle$x, ad.circle$y, type = "l")

#add labels
text(x = 0.72, y = 0.12, labels = "All genes = N", adj = 0)
text(x = 0.25, y = 0.76, labels = "DEG = K")
text(x = 0.83, y = 0.76, labels = "AD genes = n", adj = 1)
text(x = 0.5, y = 0.5, labels = "AD and DEG = k")
mtext("What is the probability that in our draw of AD genes\n
  we get k that are differentially expressed?",
  side = 1, line = -2.5, padj = 0)
```

```{r ad_enrich}
#from: https://montilab.github.io/BS831/articles/docs/HyperEnrichment.html

all.gene.id <- gsub(".value", "", rownames(full.model.result$p_values))
all.gene.name <- gene.info[match(all.gene.id, gene.info[,"ensembl_gene_id"]), "external_gene_name"]
sig.deg <- rownames(sig.expr)
#ad.orthos

p.of.draw <- phyper(q = length(ad.sig)-1, #number of red marbles in our draw (number of AD genes that are DEG)
  m = length(sig.deg), #number of red marbles (DEG) in the urn
  n = length(all.gene.id) - length(sig.deg), #number of green marbles (not DEG)
  k = length(ad.orthos), #number of drawn marbles (AD genes)
  lower.tail = FALSE #compute P(X > 1 overlap), hence the -1 above.
)
```

The calculated probability of there `r length(ad.sig)` differentially
expressed AD genes without enrichment in the set of all DEGs is 
`r signif(p.of.draw, 2)`, which we correct to `r threshold_p(p.of.draw)`.



## Testing individual factors 
we test everything in one model now, and these variables are confusing me.

We also tested the effects of each factor after adjusting
for the other factors to get effect sizes for each.
This allows us to make volcano plots for each factor.

```{r diff_fun}

test_one_var <- function(expr.mat, covar.mat, covar.names, test.var){

  var.test.file <- file.path(results.dir, paste0(expression.type, "_Effects_of_", test.var, ".RDS"))

  adj.covar <- setdiff(covar.names, test.var)
  no.var.covar <- dummy_covar(covar.mat[,adj.covar,drop=FALSE])
  adj.expr.file <- file.path(processed.data.dir, paste0("Adjusted_", expression.type, "_Expression_for_", test.var, "_test.RDS"))
  if(file.exists(adj.expr.file)){
    adj.expr <- readRDS(adj.expr.file)
  }else{
    adj.expr <- adjust(t(expr.mat), no.var.covar)
    saveRDS(adj.expr, adj.expr.file)
  }

  #match covariates to adjusted expression matrix
  common.ind <- intersect(rownames(adj.expr), rownames(covar.mat))
  expr.idx <- match(common.ind, rownames(adj.expr))
  covar.idx <- match(common.ind, rownames(covar.mat))

  diff.test <- apply(adj.expr[expr.idx,], 2, function(x) lm(x~covar.mat[covar.idx,test.var]))
  test.coef <- t(sapply(diff.test, function(x) coef(x)))
  test.p <- t(sapply(diff.test, function(x) summary(x)$coefficients[,"Pr(>|t|)"]))
  colnames(test.coef) <- colnames(test.p) <- gsub("covar.mat[, test.var]", "", colnames(test.coef), fixed = TRUE)
  test.f <- lapply(diff.test, function(x) summary(x)$fstatistic)
  overall.p <- sapply(test.f, function(x) pf(x[1],x[2],x[3],lower.tail=F))
  test.r2 <- sapply(diff.test, function(x) summary(x)$adj.r.squared)
  result <- list("coef" = test.coef, "coef_p_values" = test.p, "overall_p" = overall.p, "R2" = test.r2)
  saveRDS(result, var.test.file)

  return(result)
}
```

```{r run_diff_test}
all.tests <- vector(mode = "list", length = (length(use.covar)-1))
names(all.tests)  <- use.covar[2:length(use.covar)]

for(i in 2:length(use.covar)){
  result.idx <- which(names(all.tests) == use.covar[i])
  all.tests[[result.idx]] <- test_one_var(expr.mat = scaled.expr, 
    covar.mat = plotting.df, covar.names = use.covar, 
    test.var = use.covar[i])
}
```

### Volcano Plots {.tabset .tabset-fade .tabset-pills}

The following volcano plots show the relationship between
coefficients and p values for each of the variables tested.
Here, we used a p value of `r fdr.thresh` to mark genes that
might be differentially expressed.

Again, we removed the sex-specific genes (those with a 
-log10(p) > `r sex.effect.threshold`).

```{r volcano, results = "asis", fig.width = 8, fig.height = 4}

#pdf("~/Desktop/volcano.pdf", width = 8, height = 4)
#par(mfrow = c(3,2))
de.list <- vector(mode = "list", length = length(all.tests))
names(de.list)  <- names(all.tests)

par(mfrow = c(1,2))
for(i in 1:length(all.tests)){
  cat("####", names(all.tests)[i], "\n")
  p <- all.tests[[i]]$overall_p
  coef.p <- all.tests[[i]]$coef_p_values
  coef.table <- all.tests[[i]]$coef[,2:ncol(all.tests[[i]]$coef),drop=FALSE]

  if(names(all.tests)[i] == "sex_ge"){
    #take out the genes with the largest sex effect. These are all X and Y genes.
    big.sex.effect <- which(-log10(p) > sex.effect.threshold)
    big.sex.genes <- gene.info[match(gsub(".value", "", names(big.sex.effect)), gene.info[,"ensembl_gene_id"]),]
    p <- p[-big.sex.effect]
    coef.p <- coef.p[-big.sex.effect,,drop=FALSE]
    coef.table <- coef.table[-big.sex.effect,,drop=FALSE]
  }

  adj.p <- p.adjust(p, "fdr")
  sig.idx <- which(adj.p <= fdr.thresh)
  de.list[[i]] <- sig.idx
  #length(sig.idx)

  qqunif.plot(p,  
    plot.label = paste("qq plot for", names(all.tests)[i], 
    "\n", length(sig.idx), "significant genes"))

  sig.col <- rep("black", length(adj.p))
  sig.col[sig.idx] <- "red"
  coef.names <- gsub("info[, test.var]", "", colnames(coef.table), fixed = TRUE)
  colnames(coef.table) <- coef.names  

  #par(mfrow = c(2,2))
  for(cv in 1:ncol(coef.table)){  
    plot(coef.table[,cv], -log10(coef.p[,(cv+1)]), xlab = "Coefficient",
      ylab = "-log10(p value)", 
      main = paste(names(all.tests)[i], 
        gsub("covar.mat[covar.idx, test.var]", "", colnames(coef.table)[cv], fixed = TRUE), sep = ", "), 
      col = sig.col, pch = 16, cex = 0.7)
  }

  if(expression.type == "gene"){
    gene.id <- rownames(scaled.expr)[sig.idx]
    transcript.id <- sub.id[match(gene.id, sub.id[,1]),2]
  }
  if(expression.type == "transcript"){
    transcript.id <- rownames(scaled.expr)[sig.idx]
    gene.id <- sub.id[match(transcript.id, sub.id[,1]),2]
  }
  gene.idx <- match(gene.id, gene.info[, "ensembl_gene_id"])
  sig.gene.info <- gene.info[gene.idx,c("external_gene_name", "chromosome_name", "start_position")]
  sig.table <- cbind("geneID" = gene.id, "transcriptID" = transcript.id,
    sig.gene.info,
    "coef" = coef.table[sig.idx,],
    "p" = as.numeric(p[sig.idx]), "p_adj" = as.numeric(adj.p[sig.idx]))

  ordered.table <- sig.table[order(sig.table[,"p_adj"], decreasing = FALSE),,drop=FALSE]
  sig.gene.file <- file.path(results.dir, paste0("Significant_", expression.type, "_", 
    names(all.tests)[i], ".txt"))
  write.table(ordered.table, sig.gene.file, sep = "\t", quote = FALSE, row.names = FALSE)
  cat("\n\n")
}
#dev.off()

#num.de <- sapply(de.list, length)
#write.table(num.de, here("Results", "for_paper", paste0("num_de_", basename(results.dir), ".txt")),
#  quote = FALSE, sep = "\t", col.names = FALSE)
```

## plotting and clustering differentially expressed AD genes

This looks basically identical to the plots of the overall
DEG, so it is not worth doing over in the ms.

These expression patterns fall into two distinct groups:

```{r ad_decomp}
ad.decomp <- plot.decomp(ad.sig.mat, plot.results = FALSE)
ad.cl <- pam(ad.decomp$u, k = 2)$clustering
plot.decomp(ad.sig.mat, cols = ad.cl)
```

```{r ad_sig_gene_figure}
if(subgroup[[1]] == 12){
  #make sure FC up is the first cluster to match the general DEG results
  alt.cl <- ad.cl
  alt.cl[which(ad.cl == 1)] <- 2
  alt.cl[which(ad.cl == 2)] <- 1
  ad.cl <- alt.cl 

  geno.order <- c(3,2,4,5)
  cl.order <- rev(order(ad.cl))
  ad.plot.mat <- ad.sig.mat[cl.order,geno.order]
  scaled.mat <- t(apply(ad.plot.mat, 1, scale))
  dimnames(scaled.mat) <- dimnames(ad.plot.mat)

  png(here("Results", "for_paper", "ad_clusters.png"), width = 7, height = 3,  units = "in", res = 300)
  layout.mat <- matrix(c(1,2), nrow = 1)
  layout(layout.mat, widths = c(1, 0.2))
  par(mar = c(2,4,4,1), xpd = NA)
  imageWithText(t(scaled.mat), 
    show.text = FALSE, use.pheatmap.colors = TRUE,
    col.names = NULL, row.text.shift = 0.07, row.text.adj = 0.5)
  plot.dim <- par("usr")
  plot.height <- plot.dim[4] - plot.dim[3]
  plot.width <- plot.dim[2] - plot.dim[1]
  c1.mid <- mean(which(ad.cl[col.order] == 1))
  c2.mid <- mean(which(ad.cl[col.order] == 2))
  label.y <- (plot.dim[4]+(plot.height*0.08))
  text(x = c1.mid, y = label.y, labels = "Neuron and Synapse")
  text(x = c2.mid, y = label.y, labels = "Mitochondria and Ribosome")

  #add colored labels for the genotypes
  y.vals <- 1:4
  y.shift <- plot.height*0.06
  label.xmin <- plot.dim[1] - (plot.width*0.09)
  label.xmax <- plot.dim[1] + (plot.width*0.03)
  for(i in 1:length(y.vals)){
    draw.rectangle(label.xmin, label.xmax, (y.vals[i] - y.shift), (y.vals[i] + y.shift),
      border = geno.cols[match(rev(colnames(ad.plot.mat))[i], names(geno.cols))], 
      lwd = 3)
  }

  #add scale bar
  par(mar = c(2,4,4,1))
  imageWithTextColorbar(scaled.mat, use.pheatmap.colors = TRUE, cex = 1, bar.lwd = 3)
  dev.off()
}
```


The mean expression of the genes in each group
is shown below. The rows are scaled to accentuate
the cross-genotype differences.

```{r cl1}
cl1.idx <- which(ad.cl == 1)
pheatmap(ad.sig.mat[cl1.idx,ordered.geno], cluster_cols = FALSE, 
  scale = "row", show_rownames = FALSE)
```

The enrichments for this cluster are shown below.

```{r cl1_enrich}
cl1.enrich <- gost(rownames(ad.sig.mat)[cl1.idx], organism = "mmusculus",
  source = c("GO", "KEGG", "CORUM", "REACTOME", "HP"))
plot.enrichment(cl1.enrich, num.terms = 25, plot.label = "Cluster 1 enrichment")
```

The mean expression of the genes in the second group is
shown below.

```{r cl2}
cl2.idx <- which(ad.cl == 2)
pheatmap(ad.sig.mat[cl2.idx,ordered.geno], cluster_cols = FALSE, 
  scale = "row", show_rownames = FALSE)
```

These enrichments match the clusters seen in the differentially 
expressed genes overall, and suggest that the differential expression
caused by the variants is related to AD pathology.

```{r cl2_enrich}
cl2.enrich <- gost(rownames(ad.sig.mat)[cl2.idx], organism = "mmusculus",
  source = c("GO", "KEGG", "CORUM", "REACTOME", "HP"))
plot.enrichment(cl2.enrich, num.terms = 25, plot.label = "Cluster 2 enrichment")
```


```{r wgcna, eval = FALSE}
#I checked WGCNA. Unfortunately, it didn't really yield much.

library(WGCNA)

enableWGCNAThreads()

powers = c(c(1:10), seq(from = 12, to=20, by=2))
sft = pickSoftThreshold(adj_expr, powerVector = powers, verbose = 5)
use.power <- sft$powerEstimate
if(is.na(use.power)){use.power = 6}
use.power

#signed if no negative values
#unsigned if negative values are included
#type = "unsigned"
type = "signed"
net = blockwiseModules(adj_expr, power = use.power, 
    TOMType = type, minModuleSize = 10, 
    reassignThreshold = 0, mergeCutHeight = 0.25, 
    numericLabels = TRUE, pamRespectsDendro = FALSE, 
    saveTOMs = FALSE,verbose = 3)

mergedColors = labels2colors(net$colors)
modules <- net$colors
num.wgcna.modules <- length(unique(modules))

u_mods <- sort(unique(modules))
mod.size <- sapply(u_mods, function(x) length(which(modules == x)))
barplot(mod.size)


enrich.file <- file.path(results.dir, paste0("WGCNA_Module_Enrichment_", type, ".RDS"))
if(!file.exists(enrich.file)){
    mod.enrich <- lapply(u_mods, function(x) gost(names(which(modules == x)), 
        organism = "mmusculus", sources = c("GO", "KEGG", "REACTOME", "CORUM", "HP")))
    saveRDS(mod.enrich, enrich.file)
}else{
    mod.enrich <- readRDS(enrich.file)
}

pdf("~/Desktop/enrichment.pdf", width = 9, height = 40)
plot.enrichment.group(mod.enrich, max.term.size = 3000, transformation = "sqrt")
dev.off()

wgcna.list <- lapply(u_mods, function(x) names(which(modules == x)))
gene.r2 <- vector(mode = "list", length = length(wgcna.list))
mod.r2 <- rep(NA, length(wgcna.list))

for(i in 1:length(u_mods)){
    print(i)
    ind.genes <- lapply(wgcna.list[[i]], function(x) x)
    wgcna.vals <- non_nested_mean_vals(ind.genes, adj_expr, 
        group.representation = group.representation)
    rownames(wgcna.vals) <- unlist(ind.genes)
    wgcna.result <- plot_mean_bd_k(wgcna.vals, covar.mat = covar.table, plot.results = FALSE)
    gene.r2[[i]] <- wgcna.result$row.r2
    mod.r2[i] <- wgcna.result$overall.r2
}

barplot(sort(mod.r2))

mod.order <- order(mod.r2, decreasing = TRUE)
i = 1
ind.result <- plot_individual_genes(gene.list = names(which(modules == u_mods[mod.order[i]])))

#quartz()
plot.enrichment(mod.enrich[[u_mods[high.pred[i]]]], max.term.size = 3000)


hist(unlist(gene.r2))

sorted.genes <- sort(unlist(gene.r2), decreasing = TRUE)
plot(sorted.genes)


#This gives basically the same results to other analyses.
library(fgsea)
#gsea.enrich <- fgsea::fgsea(kegg.list, sorted.genes, scoreType = "pos")
#gsea.enrich <- fgsea::fgsea(bd.list, sorted.genes, scoreType = "pos")
gsea.enrich <- fgsea::fgsea(unlist(kegg.bd.list, recursive = FALSE), sorted.genes, scoreType = "pos")


norm.es <- as.numeric(as.matrix(gsea.enrich[,"NES"]))
names(norm.es) <- gsea.enrich$pathway

par(mar = c(4,24, 2,2))
barplot(tail(sort(norm.es), 19), las = 2, horiz = TRUE)
```

## Latent Pathway Identification Analysis (LPIA)

Here we use a method developed in Pham et al. 2011
called Latent Pathway Identification Analyais (LPIA).
This is a network-based method that combines expression
data with GO-KEGG intersections into a network in which
an edge exists if there are genes in the GO-KEGG intersection.
The edge is weighted by the product of two terms: (1) the 
Jaccard index of the two terms (in terms of genes), and 
(2) the differential expression of the genes in the overlapping
genes. In Pam et al. they were comparing two groups, so they 
used the median differential expression. We could compare the
FC and VS carriers directly to do this, or we can use a different
term that is based on our linear model with ordered groups.

They then project this bipartite network onto the the KEGG 
terms. The nodes in this network are linked if they share 
at least one biological process. The edge weights are larger
if the two KEGG terms relate more GO terms. We would maybe
use the Biodomain projection? We can try both. 

They use eigenvector centrality to calculate centrality for 
each pathway.

```{r lpia}
#sub.vals
intersect.result = sub.val.result
intersect.jaccard = kegg.bd.jaccard

#generate a network between all Biodomains and KEGG pathways
#that combines the Jaccard index of the pathways and the 
#degree to which the intersection is differentially expressed.

lpia_adj <- function(intersect.result, intersect.jaccard){
    net.adj <- matrix(0, nrow = nrow(intersect.jaccard), ncol = ncol(intersect.jaccard))
    dimnames(net.adj) <- dimnames(intersect.jaccard)
    split.name <- strsplit(names(intersect.result$row.r2), " : ")
    row.kegg <- sapply(split.name, function(x) x[1])
    row.bd <- sapply(split.name, function(x) x[2])

    for(bd in 1:nrow(intersect.jaccard)){
        bd.idx <- which(row.bd == rownames(net.adj)[bd])
        kegg.idx <- match(row.kegg[bd.idx], colnames(net.adj))
        intersect.r2 <- intersect.result$row.r2[bd.idx]
        #edge weight is the R2 times the jaccard indes
        eweight <- intersect.r2*intersect.jaccard[bd,kegg.idx]
        net.adj[bd, kegg.idx] <- eweight
    }
    return(net.adj)
}

#create a biadjacency matrix in which the edge weights 
#are a product of the KEGG-Biodomain Jaccard index and the
#R2 of the effect of genotoype on the intersection (we might
#want to change this to better capture effect size).
lpia_mat <- lpia_adj(sub.val.result, kegg.bd.jaccard)
lpia_mat[which(lpia_mat < 0)] <- 0 #zero out any negative numbers
lpia_net <- graph_from_biadjacency_matrix(lpia_mat)
lpia_net <- delete_vertices(lpia_net, which(degree(lpia_net) == 0))
#plot(lpia_net)

lpia_proj <- bipartite_projection(lpia_net)


bd.proj <- lpia_proj[[1]]
bd.proj <- delete_vertices(bd.proj, which(degree(bd.proj) == 0))

kegg.proj <- lpia_proj[[2]]
kegg.proj <- delete_vertices(kegg.proj, which(degree(kegg.proj) == 0))


#plot(bd.proj)
#plot(kegg.proj)
bd_cent <- eigen_centrality(bd.proj)
kegg_cent <- eigen_centrality(kegg.proj)

barplot(bd_cent$vector, las = 2)
par(mar = c(4,16,4,2))
barplot(tail(sort(kegg_cent$vector), 10), las = 2, horiz = TRUE)
```

## From smaller intersection analysis

## KEGG pathways with altered intersections {.tabset .tabset-fade .tabset-pills}

We then looked at all the KEGG intersections with the 
APP Processing biodomain. Because we've already calculated
all the significant correlations in the APP biodomain, we
can just look for where these fall in the overlapping 
KEGG pathways.

```{r all_intersections}
app.list <- kegg.bd.list$"APP Metabolism"
app.kegg.sig <- vector(mode = "list", length = length(app.list))
names(app.kegg.sig) <- names(app.list)
for(i in 2:length(app.list)){
    overlapping.genes <- app.list[[i]]
    overlapping.names <- mouse.genes[match(overlapping.genes, mouse.genes[,"ensembl_gene_id"]),"external_gene_name"]
    common.names <- intersect(overlapping.names, colnames(corr_result[[1]]))
    sub.net <- corr_result[[1]][common.names, common.names]
    sub.p <- corr_result[[2]][common.names, common.names]
    sig.sub <- which(sub.p <= 0.001, arr.ind = TRUE)
    if(nrow(sig.sub) > 0){
        sig.sub.names <- cbind(colnames(sub.net)[sig.sub[,1]], colnames(sub.net)[sig.sub[,2]])
        sig.sub.result <- cbind(sig.sub.names, sub.net[sig.sub], sub.p[sig.sub])
        colnames(sig.sub.result) <- c("gene1", "gene2", "effect", "p")
        app.kegg.sig[[i]] <- sig.sub.result
    }
}
has.vals <- which(sapply(app.kegg.sig, length) > 0)
sig.app.k.list <- app.kegg.sig[has.vals]
```

There were `r length(app.kegg.sig)` KEGG pathways with genes overlapping
the APP processing biodomain. Of these `r length(sig.app.k.list)` had
gene-gene correlations that were significantly dependent on Klotho genotype.

The bar plot below shows how many gene pairs in each term had altered
correlations across genotypes.

```{r num_alt}
num.pairs <- sapply(sig.app.k.list, nrow)
par(mar = c(4,22,4,2))
barplot_with_num(sort(num.pairs), horiz = TRUE, las = 2, xlab = "Number of Gene Pairs",
    main = "KEGG pathways in APP Metabolism\nwith altered gene-gene correlations")
```

```{r plot_kegg, results = "asis", fig.width = 10, fig.height = 10}
#pdf("~/Desktop/kegg_overlaps.pdf", width = 10, height = 10)
for(i in 1:length(sig.app.k.list)){
    cat("###", names(sig.app.k.list)[i], "\n")
    plot_net(dir.net, layout.mat = net.layout, plot.label = names(sig.app.k.list)[i])
    common.genes <- unique(as.vector(sig.app.k.list[[i]][,1:2]))
    common.idx <- match(common.genes, rownames(net.layout))
    scaled.idx <- net.layout
    scaled.idx[,1] <- scale.between.vals(net.layout[,1], -1, 1)
    scaled.idx[,2] <- scale.between.vals(net.layout[,2], -1, 1)
    points(scaled.idx[common.idx,1], scaled.idx[common.idx,2], cex = 2, col = "red", pch = 16)
    cat("\n\n")
}
#dev.off()
```


## Latent Pathway Identification Analysis (LPIA)

Here we use a method developed in Pham et al. 2011
called Latent Pathway Identification Analyais (LPIA).
This is a network-based method that combines expression
data with GO-KEGG intersections into a network in which
an edge exists if there are genes in the GO-KEGG intersection.
The edge is weighted by the product of two terms: (1) the 
Jaccard index of the two terms (in terms of genes), and 
(2) the differential expression of the genes in the overlapping
genes. In Pam et al. they were comparing two groups, so they 
used the median differential expression. We could compare the
FC and VS carriers directly to do this, or we can use a different
term that is based on our linear model with ordered groups.

They then project this bipartite network onto the the KEGG 
terms. The nodes in this network are linked if they share 
at least one biological process. The edge weights are larger
if the two KEGG terms relate more GO terms. We would maybe
use the Biodomain projection? We can try both. 

They use eigenvector centrality to calculate centrality for 
each pathway.

Below, we generate lists of KEGG pathways and Biodomain 
intersections and calculate Jaccard indices for all 
intersections. 

We take up the rest of the analysis in 1.KEGG_and_Biodomains.Rmd
for historical reasons.

```{r intersect}

intersect_bd_kegg <- function(bd.list, kegg.list){

  intersect.list <- vector(mode = "list", length = length(bd.list))
  names(intersect.list) <- names(bd.list)

  for(bd in 1:length(bd.list)){
    k.intersections <- vector(mode = "list", length = length(kegg.list))
    names(k.intersections) <- names(kegg.list)  
    
    for(k in 1:length(kegg.list)){
      k.intersections[[k]] <- intersect(bd.list[[bd]], kegg.list[[k]])
    }

    has.genes <- which(sapply(k.intersections, length) >= min.term.size)
    intersect.list[[bd]] <- k.intersections[has.genes]
  }
  return(intersect.list)
}

bd_kegg_jaccard <- function(bd.list, kegg.list){

  jacc.mat <- matrix(NA, nrow = length(bd.list), ncol = length(kegg.list))
  rownames(jacc.mat) <- names(bd.list)
  colnames(jacc.mat) <- names(kegg.list)

  for(bd in 1:length(bd.list)){
    k.intersections <- vector(mode = "list", length = length(kegg.list))
    names(k.intersections) <- names(kegg.list)  
    
    for(k in 1:length(kegg.list)){
      jacc.mat[bd,k] <- jaccard.ind(bd.list[[bd]], kegg.list[[k]])
    }

  }
  return(jacc.mat)
}


#save grouped form, so we can do the boxplot thingy that Greg Cary does.
mouse.bd.k.intersections <- intersect_bd_kegg(bd.list = bd.for.gsea, kegg.list = mouse.path.gene.ensembl)
mouse.bd.k.jacc <- bd_kegg_jaccard(bd.list = bd.for.gsea, kegg.list = mouse.path.gene.ensembl)
saveRDS(mouse.bd.k.intersections, file.path(general.processed.data.dir, 
  paste0("Mouse_KEGG_Intersections_for_GSEA_", expression.type, ".RDS")))
saveRDS(mouse.bd.k.jacc, file.path(general.processed.data.dir, 
  paste0("Mouse_KEGG_Intersection_Jaccard_", expression.type, ".RDS")))

human.bd.k.intersections <- intersect_bd_kegg(bd_genes, hum.path.gene.ensembl)
human.bd.k.jacc <- bd_kegg_jaccard(bd_genes, hum.path.gene.ensembl)
saveRDS(human.bd.k.intersections, file.path(general.processed.data.dir, 
  paste0("Human_KEGG_Intersections_for_GSEA_", expression.type, ".RDS")))
saveRDS(human.bd.k.jacc, file.path(general.processed.data.dir, 
  paste0("Human_KEGG_Intersections_Jaccard_", expression.type, ".RDS")))
```


```{r intersect_vals}

mean.bd.k.file <- file.path(results.dir, paste0("KEGG_BD_mean_expression_", group.representation, ".RDS"))

if(!file.exists(mean.bd.k.file)){
  mean.bd.k.vals <- vector(mode = "list", length = length(mouse.path.gene.ensembl))
  names(mean.bd.k.vals) <- names(mouse.path.gene.ensembl)

  #for each kegg pathway
  for(k in 1:length(mouse.path.gene.ensembl)){
    kegg.path <- names(mouse.path.gene.ensembl)[k]
    has.intersection <- which(sapply(lapply(mouse.bd.k.intersections, function(x) which(names(x) == kegg.path)), length) > 0)
    if(length(has.intersection) == 0){next()}

    intersect.mean.vals <- vector(mode = "list", length = length(has.intersection))
    names(intersect.mean.vals) <- names(has.intersection)

    all.val.mat <- matrix(NA, nrow = length(mouse.bd.genes), ncol = 5)
    rownames(all.val.mat) <- names(mouse.bd.genes)
    colnames(all.val.mat) <- genotypes[1:5]

    #pull out all gene values for intersections with biodomains
    for(bd in 1:length(has.intersection)){
      kegg.idx <- sapply(mouse.bd.k.intersections[has.intersection], function(x) which(names(x) == kegg.path))
      
      intersect.ids <- mouse.bd.k.intersections[[has.intersection[bd]]][[kegg.idx[bd]]]
      bd.k.vals <- lapply(intersect.ids, 
        function(x) plot_tx_with_genotype(expr.mat = scaled.expr, covar.table = covar.mat, 
        tx_name = x, tx_label = NA, ylab = "Expression (A.U.)", plot.results = FALSE)[[1]])

      has.vals <- which(sapply(bd.k.vals, length) > 0)
      k.mean <- t(sapply(bd.k.vals[has.vals], function(x) sapply(x[[1]], mean)))
      #pheatmap(k.mean)
      rownames(k.mean) <- intersect.ids[has.vals]
      intersect.mean.vals[[bd]] <- k.mean
    }
    has.vals <- which(sapply(intersect.mean.vals, length) > 0)
    if(group.representation == "means"){
      mean.mat <- t(sapply(intersect.mean.vals[has.vals], colMeans))
    }
    if(group.representation == "eigenegene"){
      mean.mat <- t(sapply(intersect.mean.vals[has.vals], function(x) plot.decomp(x, plot.results = FALSE)$v[,1]))
    }
    all.val.mat[rownames(mean.mat), colnames(mean.mat)] <- mean.mat[rownames(mean.mat), colnames(mean.mat)]
    mean.bd.k.vals[[k]] <- all.val.mat
    #pheatmap(mean.bd.k.vals[[k]], main = names(path.gene.ensembl)[k], cluster_cols = FALSE, cluster_rows = FALSE)
  }
  saveRDS(mean.bd.k.vals, mean.bd.k.file)
}else{
  mean.bd.k.vals <- readRDS(mean.bd.k.file)
}

has.vals <- which(sapply(mean.bd.k.vals, length) > 0)
min.val <- min(unlist(mean.bd.k.vals[has.vals]), na.rm = TRUE)
max.val <- max(unlist(mean.bd.k.vals[has.vals]), na.rm = TRUE)
#custom.dist <- unlist(mean.bd.k.vals[has.vals])
custom.dist <- rnorm(1000, 0, 0.2)
#hist(custom.dist)

kegg.plot.dir <- file.path(results.dir, "KEGG_plots")
if(!file.exists(kegg.plot.dir)){dir.create(kegg.plot.dir)}


pdf(file.path(kegg.plot.dir, "kegg_bd_intersections.pdf"), width = 7, height = 7)
hist(unlist(mean.bd.k.vals), xlab = "Mean Value", main = "Mean Biodomain-Kegg Intersection Expression")

for(i in 1:length(mean.bd.k.vals)){
  if(length(mean.bd.k.vals[[i]]) > 0){
    par(mar = c(4,12,2,2))
    neg.vals <- length(which(mean.bd.k.vals[[i]] < 0))
    pos.vals <- length(which(mean.bd.k.vals[[i]] > 0))
    if(neg.vals > 0 && pos.vals > 0){
      col.scale = c("blue", "brown")
      grad.dir = "ends"
      split.at.vals = TRUE
    }
    if(neg.vals > 0 && pos.vals == 0){
      col.scale = "blue"
      grad.dir = "low"
      split.at.vals = TRUE
    }
    if(neg.vals == 0 && pos.vals > 0){
      col.scale = "brown"
      grad.dir = "high"
      split.at.vals = TRUE
    }

  imageWithText(mat = mean.bd.k.vals[[i]], main = names(mean.bd.k.vals)[i], 
      col.scale = col.scale, grad.dir = grad.dir, col.text.rotation = 0, 
      col.text.adj = 0.5, col.text.shift = 0.06, row.text.shift = 0.15, 
      main.shift = 0.05, color.fun = "linear", split.at.vals = split.at.vals,
      light.dark = "f", global.color.scale = TRUE, global.min = min.val, 
      global.max = max.val)

  }
}
dev.off()

global.min <- min(unlist(mean.bd.k.vals), na.rm = TRUE)
global.max <- max(unlist(mean.bd.k.vals), na.rm = TRUE)
#plot all results for each biodomain on one page
for(bd in 1:length(bd_genes)){
  bd.mat <- t(sapply(mean.bd.k.vals, function(x) if(length(x) > 0){x[bd,]}else{rep(NA, 5)}))
  has.vals <- which(!is.na(rowMeans(bd.mat)))

  #plot.decomp(t(bd.mat[has.vals,]), label.points = TRUE, main = names(bd_genes)[bd])
  #bd.decomp <- plot.decomp(bd.mat[has.vals,], main = names(bd_genes)[bd])
  #plot.decomp(bd.mat[has.vals,], main = names(bd_genes)[bd], label.points = TRUE)

  if(length(has.vals) > 0){
    png(file.path(kegg.plot.dir, paste0("bd_kegg_intersections_", names(bd_genes)[bd], ".png")), 
      width = 7, height = max(c(length(has.vals)/7, 7)), units = "in", res = 200)
    row.order <- hclust(dist(bd.mat[has.vals,]))$order
    par(mar = c(4,16, 2, 2))
    test <- imageWithText(mat = bd.mat[has.vals[row.order],], main = names(bd_genes)[bd], 
        col.scale = c("blue", "brown"), grad.dir = "ends", col.text.rotation = 0, 
        col.text.adj = 0.5, col.text.shift = 0.01, row.text.shift = 0.15, 
        main.shift = 0.01, color.fun = "linear", split.at.vals = TRUE,
        color.dist = NULL, light.dark = "f", global.color.scale = TRUE,
        global.min = global.min, global.max = global.max)
    dev.off()
  }
}
```


We put all the results together to see if there were any combinations
that stood out overall. The plot below shows that the expression across
all KEGG pathway and biodomain intersections separates the genotypes
nicely. Along the first principal component, the VS genotypes are negative,
the WT mice are at 0, and the FC genotypes are positive.

```{r all_kegg, fig.width = 5, fig.height = 5}

big.mat <- Reduce("rbind", mean.bd.k.vals)
group.names <- unlist(lapply(1:length(mean.bd.k.vals), function(x) if(length(mean.bd.k.vals[[x]] > 0)){paste(names(mean.bd.k.vals)[x], rownames(mean.bd.k.vals[[x]]), sep = "--")}))
rownames(big.mat) <- group.names
#pdf("~/Desktop/big.mat.pdf", height = 20, width = 10)
#pheatmap(big.mat, cluster_rows = FALSE, cluster_cols = FALSE)
#dev.off()
not.na <- which(!is.na(rowSums(big.mat)))

sub.mat <- big.mat[not.na,]
process.decomp <- plot.decomp(t(sub.mat), label.points = TRUE, xlim = c(-1, 1), 
  cols = geno.cols[match(names(geno.cols), colnames(sub.mat))], cex = 1.5)
process.vals <- process.decomp$v

pc.order <- order(process.decomp$u[,1])
barplot(process.decomp$u[pc.order,1], 
  col = geno.cols[match(names(geno.cols), colnames(sub.mat))][pc.order], 
  names = colnames(sub.mat)[pc.order], ylab = "PC1")
abline(h = 0)
#process.decomp <- plot.decomp(big.mat[has.vals,], label.points = TRUE)
#plot(process.decomp$v, type = "n")
#text(process.decomp$v[,1], process.decomp$v[,2], labels = colnames(big.mat))
#process.vals <- process.decomp$u

```

The processes don't have any particular clustering as
shown below. But there are some pathways that have 
large negative and positive values in the first PC.

```{r process_pc, fig.width = 8, fig.height = 7}
plot(process.decomp$v, xlab = "PC1", ylab = "PC2", pch = 16)
large.neg <- which(process.decomp$v[,1] < -0.04)
large.pos <- which(process.decomp$v[,1] > 0.05)
text(process.decomp$v[large.pos,1], process.decomp$v[large.pos,2], labels = rownames(sub.mat)[large.pos], cex = 0.5)
text(process.decomp$v[large.neg,1], process.decomp$v[large.neg,2], labels = rownames(sub.mat)[large.neg], cex = 0.5)
```

The plots below show the mean expression for each of these
processes with large PC values. 

```{r processes, fig.width = 10, fig.height = 5}
pc.order <- order(process.vals[,1], decreasing = TRUE)
most.neg <- tail(pc.order, 20)
most.pos <- head(pc.order, 20)
#most.pos <- pc.order[1620:1630] #check middle

#test.mat <- process.decomp$v[most.neg,1,drop=FALSE]
#rownames(test.mat) <-  rownames(sub.mat)[most.neg]
#test.mat[order(test.mat[,1]),]

par(mar = c(4, 24, 2, 2))
layout(matrix(c(1,2), ncol = 2), widths = c(1, 0.5))
imageWithText(sub.mat[rev(most.neg),], col.scale = c("blue", "brown"), split.at.vals = TRUE,
  grad.dir = "ends", col.text.rotation = 0, col.text.adj = 0.5, col.text.shift = 0.05, row.text.cex = 0.7,
  row.text.shift = 0.15, main = "Large Negative PC pathways")
par(mar = c(3.8, 0, 1.8, 2))
barplot(process.vals[most.neg,1], horiz = TRUE)
mtext("PC1", side = 1, line = 2.2)


#test.mat <- process.decomp$v[most.pos,1,drop=FALSE]
#rownames(test.mat) <-  rownames(sub.mat)[most.pos]
#test.mat[order(test.mat[,1]),]

par(mar = c(4, 18, 2, 0))
#row.order <- hclust(dist(sub.mat[most.pos,]))$order
layout(matrix(c(1,2), ncol = 2), widths = c(1, 0.5))
imageWithText(sub.mat[most.pos,], col.scale = c("blue", "brown"), split.at.vals = TRUE,
  grad.dir = "ends", col.text.rotation = 0, col.text.adj = 0.5, col.text.shift = 0.05, row.text.cex = 0.7,
  row.text.shift = 0.15, main = "Large Positive PC pathways")
par(mar = c(3.8, 0, 1.8, 2))
barplot(process.vals[rev(most.pos),1], horiz = TRUE)
mtext("PC1", side = 1, line = 2.2)
```

Can we build a biodomain network that shows us where the 
biggest differences are? The biodomains were designed such

```{r bipartite}
#This is quite derived, but let's see what happens when we collect
#the loadings on PC1 for each kegg-biodomain pair.

umat <- matrix(0, nrow = length(mouse.path.gene.ensembl), ncol = length(bd_genes))
rownames(umat) <- names(mouse.path.gene.ensembl)
colnames(umat) <- names(bd_genes)
split.names <- strsplit(rownames(sub.mat), "--")
split.kegg <- sapply(split.names, function(x) x[1])
split.bd <- sapply(split.names, function(x) x[2])
for(i in 1:nrow(process.vals)){
  umat[split.kegg[i], split.bd[i]] <- process.vals[i,1]
}

#pheatmap(umat)
no.vals <- which(rowSums(umat) == 0)
has.vals <- which(rowSums(umat) != 0)
row.order <- order(rowSums(umat[has.vals,]))
col.order <- order(colSums(umat[has.vals,]))

pdf(file.path(kegg.plot.dir, "KEGG_Umat.pdf"), width = 9, height = 40)
pheatmap(umat[has.vals[row.order],col.order], cluster_cols = TRUE, cluster_rows = TRUE)
dev.off()
```

The following image shows just the rows with strong loadings
for a given biodomain-KEGG intersection.

```{r strong_vals, fig.width = 8, fig.height = 9}

#filter to rows with the largest values
#hist(umat)
lower.thresh <- -0.04
upper.thresh <- 0.04
high.val <- which(apply(umat, 1, function(x) any(x > upper.thresh)))
low.val <- which(apply(umat, 1, function(x) any(x < lower.thresh)))
strong.vals <- union(high.val, low.val)

pheatmap(umat[strong.vals,])

#pdf("~/Desktop/sub_mat.pdf", width = 8, height = 9)
#pheatmap(umat[strong.vals,])
#dev.off()

#net <- graph_from_incidence_matrix(abs(umat[strong.vals,]))
#plot(net)

#pdf("~/Desktop/bip_plot.pdf", width = 10, height = 10)
#plot_bipartite(abs(umat[strong.vals,]))
#dev.off()
```


# excised from 2.Splicing_Variation.Rmd


## Literature search

We searched for the top differentially spliced genes in the literature
along with the terms "splicing", and "brain". The number of abstracts
retrieved for each gene is shown below. 

Synj1 (Synaptojanin1) has the largest number of associated abstracts among 
the top 10 genes. Synj1 is the major synaptic polyphosphoinositide phosphatase, 
and regulates uncoating and remobilization of synaptic vesicles during 
clathrin-mediated endocytosis (pmid: 10535736). It interacts with BIN1 
(pmid: 9195986), and APOE4-induced upregulation of SYNJ1
leads to phospholipid homeostatic dysregulation in the brain, which is associated
with cognitive decline in AD (pmid: 24052255). Downregulation of Synj1 restores 
phospholipid homeostasis and reduces levels of phosphorylated tau in vitro 
(pmid: 28231468). SYNJ1 is also associated with Parkinson's Disease 
(pmid: 23804563)

A recent paper looks at effects of SNPs, including splicing effects,
in this gene and two other endocysis-related genes (pmid: 31592138).
They predicted a number of deleterious SNPs that affected splicing
of SYNJ1. 

```{r pubmed}
library(easyPubMed)
pubmed.dir <- file.path(results.dir, "PubMed")
if(!file.exists(pubmed.dir)){dir.create(pubmed.dir, recursive = TRUE)}

search.term <- lapply(sig.gene.names, function(x) paste(x, "AND splicing AND brain"))
all.pubmed <- lapply(search.term, get_pubmed_ids)
num.papers <- unlist(lapply(all.pubmed, function(x) x$Count))
with.papers <- which(num.papers != 0)
paper.counts <- cbind(sig.gene.names[with.papers], num.papers[with.papers])
number.order <- order(as.numeric(paper.counts[,2]))

barplot_with_num(as.numeric(paper.counts[number.order,2]), 
    names = paper.counts[number.order,1], las = 2, ylab = "Number of Abstracts")

#has.data <- all.pubmed[with.papers]
#all.abstracts <- lapply(has.data, function(x) fetch_pubmed_data(x, format = "abstract"))
#for(i in 1:length(all.abstracts)){
#    write.table(all.abstracts[[i]], file.path(pubmed.dir, 
#        paste0("PubMed_Query_", tissue.names[tx], "_Up_", 
#        rownames(up.table)[with.papers[i]], ".txt")), 
#        sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
#}

```

## Exon Stats

Here we investigate some basic questions about transcript
length and relative expression in `r ref.geno` and `r alt.geno` 
animals.

```{r sandbox, eval = FALSE}
gene.name = "Synj1"
gene.id <- tx.info[match(gene.name, tx.info[,"external_gene_name"]), "ensembl_gene_id"]
plotProportions(d, gene_id = gene.id, group_variable = "genotype", 
  plot_type = "boxplot1", group_colors = geno.cols[c(1,3)])

gene.idx <- which(padj[,"geneID"] == gene.id)
tx.id <- padj[gene.idx,"txID"]
tx.idx <- match(tx.id, tx.info[,"ensembl_transcript_id"])
tx.len <- tx.info[tx.idx,"transcript_length"]
names(tx.len) <- tx.info[tx.idx,"ensembl_transcript_id"]
canonical.idx <- which(tx.info[tx.idx,"transcript_is_canonical"] == 1)
len.order <- order(tx.len)
bar.col <- rep("gray", length = length(tx.len))
bar.col[canonical.idx] <- "red"

par(mar = c(4, 12, 4, 2))
barplot(tx.len[len.order], col = bar.col[len.order], las = 2, horiz = TRUE,
  main = paste(gene.name, "transcript lengths"), xlab = "bp")

#is the canonical transcript differentially abundant?
sig.tx.idx <- which(padj[,"transcript"] <= alpha)
sig.tx.id <- padj[sig.tx.idx,"txID"]
common.tx <- intersect(sig.tx.id, tx.info[,"ensembl_transcript_id"])
is.canonical <- sapply(common.tx, function(x) 
  tx.info[match(x, tx.info[,"ensembl_transcript_id"]), "transcript_is_canonical"])
barplot_with_num(c("canonical" = length(which(is.canonical == 1)), "non-canonical" = length(which(is.na(is.canonical)))))

#what are the lengths of the differentially expressed transcripts?
tx.len <- sapply(common.tx, function(x) 
  tx.info[match(x, tx.info[,"ensembl_transcript_id"]), "transcript_length"])
#hist(tx.len)


common.idx <- match(common.tx, props[,"feature_id"])
allele.prop <- lapply(allele.names, function(x) rowMeans(as.matrix(props[common.idx,x])))
plot(allele.prop$FC, tx.len)
plot(allele.prop$VS, tx.len)
pt.col <- rep("black", length(tx.len))
pt.col[which(is.canonical == 1)] <- "red"
plot(allele.prop$FC, allele.prop$VS, col = pt.col, pch = 16)
abline(0,1)

allele.diff <- allele.prop$VS - allele.prop$FC
vs.higher <- which(allele.diff > 0)
vs.lower <- which(allele.diff < 0)

#VS has reduced expression of longer transcripts? May indicate
#intron retention in FC?
boxplot(list("up in VS" = tx.len[vs.higher], "down in VS" = tx.len[vs.lower]))

canonical.idx <- which(is.canonical == 1)
non.canonical.idx <- which(is.na(is.canonical))

plot(allele.diff, tx.len, col = pt.col, pch = 16, xlab = "Proportion difference",
  ylab = "Transcript length")
abline(v = 0)
text(x = -0.3, y = 25000, labels = "Lower in VS")
text(x = 0.2, y = 25000, labels = "Higher in VS")

#vs has slightly lower expression of canonical transcripts?
#and slightly higher of non-canonical? I don't think they
#are really different.
canon.list <- list("FC" = list("canonical" = allele.prop$FC[canonical.idx], 
  "non-canonical" = allele.prop$FC[non.canonical.idx]),
  "VS" = list("canonical" = allele.prop$VS[canonical.idx], 
  "non-canonical" = allele.prop$VS[non.canonical.idx]))
plot.grouped.boxes(canon.list, group.cols = geno.cols[c(1,3)],
  ylab = "proportion")

len.list <- list("FC" = list("canonical" = allele.prop$FC[canonical.idx], 
  "non-canonical" = allele.prop$FC[non.canonical.idx]),
  "VS" = list("canonical" = allele.prop$VS[canonical.idx], 
  "non-canonical" = allele.prop$VS[non.canonical.idx]))

```


```{r transcript_comparison}
#make a big comparison table that we can use to ask different quesitons 
#of the significant transcripts

sig.tx.idx <- which(padj[,"transcript"] <= alpha)
sig.tx.id <- padj[sig.tx.idx,"txID"]
tx.idx <- match(sig.tx.id, tx.info[,"ensembl_transcript_id"])
tx.len <- tx.info[tx.idx,"transcript_length"]
tx.canon <- tx.info[tx.idx,"transcript_is_canonical"]

alleles <- c(ref.geno, alt.geno)
props <- proportions(d)
allele.names <- lapply(alleles, 
  function(x) rownames(plotting.df)[which(plotting.df[,"ordered_geno"] == x)])
names(allele.names) <- alleles
tx.prop.idx <- match(sig.tx.id, props[,"feature_id"])
allele.prop <- sapply(allele.names, function(x) rowMeans(as.matrix(props[tx.prop.idx,x])))

#if the difference is greater than 0, the alternative allele is higher
allele.diff <- allele.prop[,alt.geno] - allele.prop[,ref.geno]
alt.higher <- which(allele.diff > 0)
alt.lower <- which(allele.diff < 0)

comp.table <- cbind("txLen" = tx.len, 
  "canonical" = tx.canon, allele.prop, "difference" = allele.diff)
rownames(comp.table) <- sig.tx.id
```

Are the transcripts that are upregulated in `r alt.geno`
longer or shorter than transcripts upregulated in `r alt.geno`?

The transcripts that are up in VS may be slightly shorter 
than those that are down in VS. Maybe this indicates more
intron retention in the FC animals?

```{r tx_len, fig.width = 5, fig.height = 5}
len.list <- list(log10(comp.table[alt.higher,"txLen"]),
  log10(comp.table[alt.lower, "txLen"]))
test <- aov.by.list(len.list)
boxplot(len.list, names = paste(c("Higher in", "Lower in"), alt.geno),
  ylab = "log10 Transcript length", main = paste("p =", signif(test$"Pr(>F)"[1], 2)))
```

Are the transcripts that are up in VS more likely to be
the canonical transcripts?



```{r canonical, fig.width = 4, fig.height = 3}
canonical.table <- matrix(NA, nrow = 2, ncol = 2)
colnames(canonical.table) <- c(c("canonical", "non-canonical"))
rownames(canonical.table) <- paste(c("Higher in", "Lower in"), alt.geno)
canonical.table[1, "canonical"] <- 
  length(intersect(alt.higher, which(comp.table[,"canonical"] == 1)))
canonical.table[1, "non-canonical"] <- 
  length(intersect(alt.higher, which(is.na(comp.table[,"canonical"]))))

canonical.table[2, "canonical"] <- 
  length(intersect(alt.lower, which(comp.table[,"canonical"] == 1)))
canonical.table[2, "non-canonical"] <- 
  length(intersect(alt.lower, which(is.na(comp.table[,"canonical"]))))

test <- fisher.test(canonical.table)
pheatmap(canonical.table, display_numbers = TRUE, cluster_rows = FALSE, 
  cluster_cols = FALSE, main = paste("p =", signif(test$p.value, 2)))
```

The plot below shows the expected counts for these categories.

```{r null, fig.width = 4, fig.height = 3}
#permute to see what the expected numbers are
nperm = 100
null.counts <- vector(mode = "list", length = nperm)
for(p in 1:nperm){
  sample.table <- comp.table
  sample.diff <- sample(comp.table[,"difference"])
  sample.higher <- which(sample.diff > 0)
  sample.lower <- which(sample.diff < 0)

  sample.counts <- matrix(NA, nrow = 2, ncol = 2)
  colnames(sample.counts) <- c(c("canonical", "non-canonical"))
  rownames(sample.counts) <- paste(c("Higher in", "Lower in"), alt.geno)
  sample.counts[1, "canonical"] <- 
    length(intersect(sample.higher, which(comp.table[,"canonical"] == 1)))
  sample.counts[1, "non-canonical"] <- 
    length(intersect(sample.higher, which(is.na(comp.table[,"canonical"]))))

  sample.counts[2, "canonical"] <- 
    length(intersect(sample.lower, which(comp.table[,"canonical"] == 1)))
  sample.counts[2, "non-canonical"] <- 
    length(intersect(sample.lower, which(is.na(comp.table[,"canonical"]))))
  null.counts[[p]] <- sample.counts
}

null.avg <- Reduce("+", null.counts)/nperm
pheatmap(null.avg, display_numbers = TRUE, cluster_rows = FALSE, cluster_cols = FALSE)
```

The plot below shows the direct comparison. I think 
there isn't a real difference in the expression of 
canonical transcripts. There are just so many transcripts
that the p value is significant.

If anything, the expression of non-canonical transcripts
is slightly higher in the VS group. 

```{r count_comp, fig.height = 5, fig.width = 5}
mat.idx <- which(canonical.table > 0, arr.ind = TRUE)
pt.names <- apply(mat.idx, 1, 
  function(x) paste(rownames(canonical.table)[x[1]], colnames(canonical.table)[x[2]]))
plot(log(canonical.table), log(null.avg), xlim = c(5.5, 7.5), ylim = c(6, 7), pch = 16)
abline(0, 1)
text(log(canonical.table), log(null.avg), labels = pt.names)
```